{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "80ef9459-980f-4ddc-acb7-6bf0e2a67578",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from src.iso_perplexity.utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7b30fe1b-19cd-4ff1-9318-5db1f6268be7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def manuscript_sample_df():\n",
    "    df = pd.DataFrame()\n",
    "    df['gene_id'] = ['A' for i in range(8*4)]\n",
    "    df['transcript_id'] = flatten_list([[f'A_{j+1}' for i in range(4)] for j in range(8)])\n",
    "    df['sample'] = flatten_list([['heart', 'brain', 'lungs', 'kidney'] for i in range(8)])\n",
    "    df['pi'] = [i*100 for i in [0.5, 0.45, 0.2, 0.2,\n",
    "                0, 0, 0.1, 0.2,\n",
    "                0, 0, 0.08, 0.2,\n",
    "                0.5, 0.1, 0.02, 0.2,\n",
    "                0, 0, 0.5, 0,\n",
    "                0, 0, 0.04, 0.2,\n",
    "                0, 0.45, 0, 0,\n",
    "                0, 0, 0.06, 0]]\n",
    "    \n",
    "    # long formitze it\n",
    "    df = df.pivot(index=['gene_id', 'transcript_id'], columns='sample', values='pi')\n",
    "    df = df.reset_index()\n",
    "    df.columns.name = ''\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5f2fafec-c4be-4ee4-9028-9c39465a3823",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>counts</th>\n",
       "      <th>tpm</th>\n",
       "      <th>pi</th>\n",
       "      <th>gene_potential</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perplexity</th>\n",
       "      <th>n_effective</th>\n",
       "      <th>feature_rank</th>\n",
       "      <th>effective</th>\n",
       "      <th>sample</th>\n",
       "      <th>n_samples_effective</th>\n",
       "      <th>expression_breadth</th>\n",
       "      <th>n_exp_samples</th>\n",
       "      <th>expression_var</th>\n",
       "      <th>avg_transcript_id_tpm</th>\n",
       "      <th>max_transcript_id_tpm</th>\n",
       "      <th>avg_gene_id_tpm</th>\n",
       "      <th>max_gene_id_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>45.0</td>\n",
       "      <td>450000.0</td>\n",
       "      <td>0.45</td>\n",
       "      <td>3</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>brain</td>\n",
       "      <td>4.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.160078</td>\n",
       "      <td>337500.0</td>\n",
       "      <td>500000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>brain</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.095743</td>\n",
       "      <td>150000.0</td>\n",
       "      <td>200000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>A_3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "      <td>brain</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.094516</td>\n",
       "      <td>140000.0</td>\n",
       "      <td>200000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A</td>\n",
       "      <td>A_4</td>\n",
       "      <td>10.0</td>\n",
       "      <td>100000.0</td>\n",
       "      <td>0.10</td>\n",
       "      <td>3</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>brain</td>\n",
       "      <td>3.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.210000</td>\n",
       "      <td>205000.0</td>\n",
       "      <td>500000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>A_5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "      <td>brain</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>500000.0</td>\n",
       "      <td>500000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id transcript_id  counts       tpm    pi  gene_potential   entropy  \\\n",
       "0       A           A_1    45.0  450000.0  0.45               3  1.368996   \n",
       "1       A           A_2     0.0       0.0  0.00               3  1.368996   \n",
       "2       A           A_3     0.0       0.0  0.00               3  1.368996   \n",
       "3       A           A_4    10.0  100000.0  0.10               3  1.368996   \n",
       "4       A           A_5     0.0       0.0  0.00               3  1.368996   \n",
       "\n",
       "   perplexity  n_effective  feature_rank  effective sample  \\\n",
       "0    2.582907          3.0             1       True  brain   \n",
       "1    2.582907          3.0             4      False  brain   \n",
       "2    2.582907          3.0             5      False  brain   \n",
       "3    2.582907          3.0             3       True  brain   \n",
       "4    2.582907          3.0             6      False  brain   \n",
       "\n",
       "   n_samples_effective  expression_breadth  n_exp_samples  expression_var  \\\n",
       "0                  4.0               100.0              4        0.160078   \n",
       "1                  2.0                50.0              4        0.095743   \n",
       "2                  2.0                50.0              4        0.094516   \n",
       "3                  3.0                75.0              4        0.210000   \n",
       "4                  1.0                25.0              4        0.250000   \n",
       "\n",
       "   avg_transcript_id_tpm  max_transcript_id_tpm  avg_gene_id_tpm  \\\n",
       "0               337500.0               500000.0        1000000.0   \n",
       "1               150000.0               200000.0        1000000.0   \n",
       "2               140000.0               200000.0        1000000.0   \n",
       "3               205000.0               500000.0        1000000.0   \n",
       "4               500000.0               500000.0        1000000.0   \n",
       "\n",
       "   max_gene_id_tpm  \n",
       "0        1000000.0  \n",
       "1        1000000.0  \n",
       "2        1000000.0  \n",
       "3        1000000.0  \n",
       "4        1000000.0  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = manuscript_sample_df()\n",
    "df = compute_multi_sample_isoform_metrics(df)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2dd49e18-d92c-49f3-8d51-c0876114583f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>perplexity</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A</td>\n",
       "      <td>4.377939</td>\n",
       "      <td>lungs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_id  perplexity  sample\n",
       "0        A    2.582907   brain\n",
       "8        A    2.000000   heart\n",
       "16       A    5.000000  kidney\n",
       "24       A    4.377939   lungs"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['gene_id', 'perplexity', 'sample']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8c989610-33ff-4ce3-9848-1a13296da176",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_potential</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A</td>\n",
       "      <td>5</td>\n",
       "      <td>kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A</td>\n",
       "      <td>7</td>\n",
       "      <td>lungs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_id  gene_potential  sample\n",
       "0        A               3   brain\n",
       "8        A               2   heart\n",
       "16       A               5  kidney\n",
       "24       A               7   lungs"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['gene_id', 'gene_potential', 'sample']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e73f52e2-8215-4aa9-993f-ce9eef67db28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>entropy</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A</td>\n",
       "      <td>2.321928</td>\n",
       "      <td>kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A</td>\n",
       "      <td>2.130252</td>\n",
       "      <td>lungs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_id   entropy  sample\n",
       "0        A  1.368996   brain\n",
       "8        A  1.000000   heart\n",
       "16       A  2.321928  kidney\n",
       "24       A  2.130252   lungs"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['gene_id', 'entropy', 'sample']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f8b651d6-5ed6-406c-9331-d335c2f23096",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>effective</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_1</td>\n",
       "      <td>True</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A_2</td>\n",
       "      <td>False</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A_3</td>\n",
       "      <td>False</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A_4</td>\n",
       "      <td>True</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A_5</td>\n",
       "      <td>False</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A_6</td>\n",
       "      <td>False</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A_7</td>\n",
       "      <td>True</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A_8</td>\n",
       "      <td>False</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A_1</td>\n",
       "      <td>True</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A_2</td>\n",
       "      <td>False</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A_3</td>\n",
       "      <td>False</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A_4</td>\n",
       "      <td>True</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A_5</td>\n",
       "      <td>False</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A_6</td>\n",
       "      <td>False</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A_7</td>\n",
       "      <td>False</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A_8</td>\n",
       "      <td>False</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A_1</td>\n",
       "      <td>True</td>\n",
       "      <td>kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A_2</td>\n",
       "      <td>True</td>\n",
       "      <td>kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>A_3</td>\n",
       "      <td>True</td>\n",
       "      <td>kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A_4</td>\n",
       "      <td>True</td>\n",
       "      <td>kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>A_5</td>\n",
       "      <td>False</td>\n",
       "      <td>kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>A_6</td>\n",
       "      <td>True</td>\n",
       "      <td>kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A_7</td>\n",
       "      <td>False</td>\n",
       "      <td>kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>A_8</td>\n",
       "      <td>False</td>\n",
       "      <td>kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A_1</td>\n",
       "      <td>True</td>\n",
       "      <td>lungs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>A_2</td>\n",
       "      <td>True</td>\n",
       "      <td>lungs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A_3</td>\n",
       "      <td>True</td>\n",
       "      <td>lungs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>A_4</td>\n",
       "      <td>False</td>\n",
       "      <td>lungs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>A_5</td>\n",
       "      <td>True</td>\n",
       "      <td>lungs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A_6</td>\n",
       "      <td>False</td>\n",
       "      <td>lungs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>A_7</td>\n",
       "      <td>False</td>\n",
       "      <td>lungs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>A_8</td>\n",
       "      <td>False</td>\n",
       "      <td>lungs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   transcript_id  effective  sample\n",
       "0            A_1       True   brain\n",
       "1            A_2      False   brain\n",
       "2            A_3      False   brain\n",
       "3            A_4       True   brain\n",
       "4            A_5      False   brain\n",
       "5            A_6      False   brain\n",
       "6            A_7       True   brain\n",
       "7            A_8      False   brain\n",
       "8            A_1       True   heart\n",
       "9            A_2      False   heart\n",
       "10           A_3      False   heart\n",
       "11           A_4       True   heart\n",
       "12           A_5      False   heart\n",
       "13           A_6      False   heart\n",
       "14           A_7      False   heart\n",
       "15           A_8      False   heart\n",
       "16           A_1       True  kidney\n",
       "17           A_2       True  kidney\n",
       "18           A_3       True  kidney\n",
       "19           A_4       True  kidney\n",
       "20           A_5      False  kidney\n",
       "21           A_6       True  kidney\n",
       "22           A_7      False  kidney\n",
       "23           A_8      False  kidney\n",
       "24           A_1       True   lungs\n",
       "25           A_2       True   lungs\n",
       "26           A_3       True   lungs\n",
       "27           A_4      False   lungs\n",
       "28           A_5       True   lungs\n",
       "29           A_6      False   lungs\n",
       "30           A_7      False   lungs\n",
       "31           A_8      False   lungs"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['transcript_id', 'effective', 'sample']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a164be9-044d-43c9-866d-edb9cfe4fd1d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0c947e5-cb8a-431a-b6e3-58d5e681b1b7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebecb882-f16a-4880-a92c-8b54169b6d28",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f1f3f8c0-d5b5-4db9-a91e-5f97237bad43",
   "metadata": {},
   "outputs": [],
   "source": [
    "def flatten_list(l):\n",
    "    \"\"\"\n",
    "    Flatten a list into 1 dimension.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    l : list\n",
    "    \"\"\"\n",
    "    return [j for i in l for j in i]\n",
    "\n",
    "def manuscript_sample_df():\n",
    "    df = pd.DataFrame()\n",
    "    df['gene_id'] = ['A' for i in range(8*4)]\n",
    "    df['transcript_id'] = flatten_list([[f'A_{j+1}' for i in range(4)] for j in range(8)])\n",
    "    df['sample'] = flatten_list([['heart', 'brain', 'lungs', 'kidney'] for i in range(8)])\n",
    "    df['pi'] = [0.5, 0.45, 0.2, 0.2,\n",
    "                0, 0, 0.1, 0.2,\n",
    "                0, 0, 0.08, 0.2,\n",
    "                0.5, 0.1, 0.02, 0.2,\n",
    "                0, 0, 0.5, 0,\n",
    "                0, 0, 0.04, 0.2,\n",
    "                0, 0.45, 0, 0,\n",
    "                0, 0, 0.06, 0]\n",
    "    \n",
    "    # long formitze it\n",
    "    df = df.pivot(index=['gene_id', 'transcript_id'], columns='sample', values='pi')\n",
    "    df = df.reset_index()\n",
    "    df.columns.name = ''\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b9bf453d-2d82-4d71-bcf5-18fd05989e97",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_global_isoform_metrics(df,\n",
    "                                   gene_col=GENE_COL,\n",
    "                                   feature_col=TRANSCRIPT_COL,\n",
    "                                   expression_col=EXP_COL,\n",
    "                                   expression_col_type='counts'):\n",
    "    \"\"\"\n",
    "    Compute isoform or other feature diversity metrics for a single-sample (bulk) dataframe.\n",
    "    Either provide counts or TPMs; if counts, will automatically convert to TPM.\n",
    "    Optionally, collapse counts to a different feature or compute TPM.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        DataFrame containing counts at transcript level.\n",
    "    gene_col : str\n",
    "        Column representing the gene (default: 'gene_id')\n",
    "    feature_col : str\n",
    "        Column representing the feature (default: 'transcript_id')\n",
    "    expression_col : str\n",
    "        Column containing expression values (default: 'counts')\n",
    "    expression_col_type : str\n",
    "        Type of expression values contained in <expression_col>\n",
    "        'counts' or 'tpm' (default: 'counts')\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        DataFrame with computed metrics.\n",
    "    \"\"\"\n",
    "    # repeatedly used column args\n",
    "    col_kwargs = {'gene_col': gene_col,\n",
    "                  'feature_col': feature_col,\n",
    "                  'expression_col': expression_col,\n",
    "                  'sample_col': None}\n",
    "\n",
    "    \n",
    "    # validate input\n",
    "    validate_counts_input(df,\n",
    "                          gene_col=gene_col,\n",
    "                          feature_col=feature_col)\n",
    "        \n",
    "    # convert from wide to long\n",
    "    df = wide_to_long(df, \n",
    "                      gene_col=gene_col,\n",
    "                      feature_col=feature_col,\n",
    "                      expression_col=expression_col)\n",
    "\n",
    "    # collapse counts if feature_col is not unique\n",
    "    df = collapse_counts_by_feature(df, **col_kwargs)\n",
    "\n",
    "    # compute TPM if requested\n",
    "    if expression_col_type == 'counts':\n",
    "        df = compute_tpm(df)\n",
    "    else:\n",
    "        df['tpm'] = df[expression_col]\n",
    "\n",
    "    # compute isoform ratios\n",
    "    df = compute_pi(df)\n",
    "    \n",
    "    # repeatedly used column args\n",
    "    col_kwargs = {'gene_col': gene_col,\n",
    "                  'feature_col': feature_col}\n",
    "    \n",
    "    # compute gene-level metrics\n",
    "    df = compute_gene_potential(df, **col_kwargs)\n",
    "    df = compute_entropy(df, gene_col=gene_col)\n",
    "    df = compute_perplexity(df)\n",
    "\n",
    "    # mark effective features\n",
    "    df = call_effective(df, **col_kwargs)\n",
    "\n",
    "    return df\n",
    "\n",
    "def compute_multi_sample_isoform_metrics(\n",
    "    df: pd.DataFrame,\n",
    "    gene_col: str = GENE_COL,\n",
    "    feature_col: str = TRANSCRIPT_COL,\n",
    "    expression_col: str = EXP_COL,\n",
    "    expression_col_type: str = 'counts'\n",
    "):\n",
    "    \"\"\"\n",
    "    Compute isoform metrics across multiple samples as well as global\n",
    "    metrics.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        Input table with counts or TPMs for all samples.\n",
    "    gene_col : str\n",
    "        Column identifying genes.\n",
    "    feature_col : str\n",
    "        Column identifying isoforms or other features (e.g. ORFs).\n",
    "    expression_col_type : {'counts', 'tpm'}\n",
    "        Type of values in `expression_col`.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        DataFrame with:\n",
    "          • per-sample metrics (gene potential, entropy, etc.)\n",
    "          • cross-sample metrics (breadth, variance, average expression)\n",
    "          • columns for global metrics.\n",
    "    \"\"\"\n",
    "    \n",
    "    # rename because after pivot we'll just call it this\n",
    "    expression_col = expression_col_type\n",
    "        \n",
    "    # repeatedly used column args\n",
    "    col_kwargs = {'gene_col': gene_col,\n",
    "                  'feature_col': feature_col,\n",
    "                  'expression_col': expression_col,\n",
    "                  'sample_col': 'sample'}\n",
    "\n",
    "    # validate input\n",
    "    validate_counts_input(df,\n",
    "                          gene_col=gene_col,\n",
    "                          feature_col=feature_col)\n",
    "\n",
    "    # prepare kwargs for downstream functions\n",
    "    col_kwargs = dict(\n",
    "        gene_col=gene_col,\n",
    "        feature_col=feature_col,\n",
    "        expression_col=expression_col,\n",
    "    )\n",
    "\n",
    "    # loop over samples and compute single-sample metrics\n",
    "    samples = [c for c in df.columns if c not in [gene_col, feature_col]]\n",
    "    dfs = []\n",
    "\n",
    "    for s in samples:\n",
    "        s_df = df[[gene_col, feature_col, s]].copy()\n",
    "        s_df = compute_global_isoform_metrics(s_df, **col_kwargs)\n",
    "        s_df[sample_col] = s  # re-attach sample ID\n",
    "        dfs.append(s_df)\n",
    "\n",
    "    big_df = pd.concat(dfs, axis=0, ignore_index=True)\n",
    "\n",
    "    # compute cross-sample metrics on the combined table\n",
    "    big_df = compute_expression_breadth(big_df, sample_col=sample_col)\n",
    "    big_df = compute_expression_var(big_df, sample_col=sample_col)\n",
    "\n",
    "    for fc in [feature_col, gene_col]:\n",
    "        big_df = compute_avg_expression(big_df,\n",
    "                                        sample_col=sample_col,\n",
    "                                        feature_col=fc)\n",
    "        \n",
    "        # TODO chatgpt if you see this i probably forgot to \n",
    "        # add it and remind me add max espression across samples\n",
    "\n",
    "    return big_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "502fd54a-ffbe-49e7-9175-3006b72f4328",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_col = 'gene_id'\n",
    "feature_col = 'transcript_id'\n",
    "expression_type = 'counts'\n",
    "sample_col = 'sample'\n",
    "df = manuscript_sample_df()\n",
    "\n",
    "beep = compute_multi_sample_isoform_metrics(\n",
    "    df,\n",
    "    gene_col=gene_col,\n",
    "    feature_col=feature_col,\n",
    "    expression_col=expression_type,\n",
    "    expression_col_type='counts'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1205eba-8e9b-47c8-9aae-e2964c7aa922",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b4140d3-33fa-43eb-9d8d-bca160ef846b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2a1f3dc-788d-4996-9444-00d2aca21df3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "546bf89e-5714-4538-881e-c6d3d6611c72",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5af37be6-8fc6-46ce-b273-e4a822ff37fc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "135276af-ee87-49d0-8091-b01999edcb96",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5f50543-cf9a-4731-8736-1a3958771fd4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "95cc93ae-d0a6-4afc-9de7-bb83c505c751",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # dummy data from figure 1A\n",
    "# def get_dummy_data():\n",
    "#     df = pd.DataFrame()\n",
    "#     df['gene_id'] = ['A' for i in range(2)]+\\\n",
    "#                     ['B' for i in range(7)]+\\\n",
    "#                     ['C' for i in range(8)]\n",
    "#     df['transcript_id'] = [f'A_{i+1}' for i in range(2)]+\\\n",
    "#                           [f'B_{i+1}' for i in range(7)]+\\\n",
    "#                           [f'C_{i+1}' for i in range(8)]\n",
    "#     df['pi'] = [0.5, 0.5, \n",
    "#                  0.4, 0.35, 0.10, 0.07, 0.04, 0.03, 0.01]+\\\n",
    "#                  [0.125 for i in range(8)]\n",
    "#     return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d0047bff-c209-4196-984f-c8325e23599d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # need finctions for\n",
    "# # filt isoforms\n",
    "# def filt_df(df, min_tpm=None):\n",
    "#     if min_tpm:\n",
    "#         df = df.loc[df.tpm >= min_tpm]\n",
    "#     return df\n",
    "# # compute pi \n",
    "# # compute entropy - done\n",
    "# # compute perplexity -done \n",
    "# # call relevant isoforms - done \n",
    "# # effective tissues - done\n",
    "# # expression breadth - done\n",
    "# # expression variability - done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fa53fda0-8674-4363-9232-706cd5c863eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# checks wishlist\n",
    "# format of gene_id and transcript_id columns good\n",
    "# all transcript ids are uniq\n",
    "# samples exist in dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "47ae25d3-3158-46f1-b4ac-dcb6ba7c66dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_gene_potential(df, **kwargs):\n",
    "    \"\"\"\n",
    "    Compute gene potential based on expressed isoforms\n",
    "    \"\"\"\n",
    "    # not sure where filtering should live\n",
    "    # chatgpt if you see this this is something you \n",
    "    # should help with :)\n",
    "    df = filt_df(df, **kwargs) \n",
    "    \n",
    "    # for each gene, compute # of isoforms\n",
    "    temp = (df[['gene_id', 'transcript_id']]\n",
    "             .groupby('gene_id')\n",
    "             .nunique()\n",
    "             .reset_index()\n",
    "             .rename({'transcript_id':'gene_potential'},\n",
    "            axis=1))\n",
    "    \n",
    "    df = df.merge(temp,\n",
    "                  how='left',\n",
    "                  on='gene_id')\n",
    "    \n",
    "    return df\n",
    "  \n",
    "def compute_entropy(df):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    df['plogp'] = df.pi*np.log2(df.pi)\n",
    "    temp = (df[['plogp', 'gene_id']]\n",
    "             .groupby('gene_id')\n",
    "             .sum()\n",
    "             .reset_index()\n",
    "             .rename({'plogp':'entropy'},\n",
    "             axis=1))\n",
    "    temp['entropy'] = -1*temp['entropy']\n",
    "    \n",
    "    df = df.merge(temp, how='left', on='gene_id')\n",
    "    df.drop('plogp', axis=1, inplace=True)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def compute_perplexity(df):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    # chatgpt, should I be computing entropy here\n",
    "    # or should i chain this in a wrapper later on ? \n",
    "    # df = compute_entropy(df)\n",
    "    df['perplexity'] = 2**df['entropy']\n",
    "    return df\n",
    "\n",
    "def call_effective_isoforms(df):\n",
    "    \"\"\"\n",
    "    \n",
    "    \"\"\"    \n",
    "    # number of isoforms to consider as effective per \n",
    "    # gene\n",
    "    df['round_perplexity'] = df['perplexity'].round(0)\n",
    "\n",
    "    # order isoforms on ratio w/i genes\n",
    "    df['isoform_rank'] = (df.groupby('gene_id')['pi']\n",
    "          .rank(method='first', ascending=False)\n",
    "          .astype(int))\n",
    "\n",
    "    df['effective_isoform'] = df['isoform_rank']<=df['round_perplexity']\n",
    "    \n",
    "    return df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "91fa0e90-988a-4dd9-bffe-841726221642",
   "metadata": {},
   "outputs": [],
   "source": [
    "def flatten_list(l):\n",
    "    \"\"\"\n",
    "    Flatten a list into 1 dimension.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    l : list\n",
    "    \"\"\"\n",
    "    return [j for i in l for j in i]\n",
    "\n",
    "def dummy_data_2():\n",
    "    df = pd.DataFrame()\n",
    "    df['gene_id'] = ['A' for i in range(8*4)]\n",
    "    df['transcript_id'] = flatten_list([[f'A_{j+1}' for i in range(4)] for j in range(8)])\n",
    "    df['sample'] = flatten_list([['heart', 'brain', 'lungs', 'kidney'] for i in range(8)])\n",
    "    df['pi'] = [0.5, 0.45, 0.2, 0.2,\n",
    "                0, 0, 0.1, 0.2,\n",
    "                0, 0, 0.08, 0.2,\n",
    "                0.5, 0.1, 0.02, 0.2,\n",
    "                0, 0, 0.5, 0,\n",
    "                0, 0, 0.04, 0.2,\n",
    "                0, 0.45, 0, 0,\n",
    "                0, 0, 0.06, 0]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "b251563e-8d85-4dd9-9d09-0a21608c4a1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute_gene_potential(df)\n",
    "# perp_df = compute_perplexity(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "873bd750-f1e9-4263-8965-1aef138dc33f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fake workflow for cross-dataset\n",
    "# calculations\n",
    "def workflow_1(df):\n",
    "    # df = df.loc[df.pi > 0] # maybe do some more sophisticated filtering here\n",
    "    df = compute_gene_potential(df)\n",
    "    df = compute_entropy(df)\n",
    "    df = compute_perplexity(df)\n",
    "    df = call_effective_isoforms(df)\n",
    "    return df\n",
    "\n",
    "# workflow 2 -- compute these metrics per sample\n",
    "def workflow_2(df, sample_col):\n",
    "    samples = df[sample_col].unique().tolist()\n",
    "    big_df = pd.DataFrame()\n",
    "    for s in samples:\n",
    "        s_df = df.loc[df[sample_col]==s].copy(deep=True)\n",
    "        s_df = workflow_1(s_df)\n",
    "        big_df = pd.concat([big_df,s_df], axis=0)\n",
    "        \n",
    "    df = big_df.copy(deep=True)\n",
    "    \n",
    "    # compute % effective tissues / isoform\n",
    "    n_samples = len(samples)\n",
    "    temp = (df.loc[df.effective_isoform][['transcript_id', sample_col]]\n",
    "            .groupby('transcript_id')\n",
    "            .nunique()\n",
    "            .reset_index()\n",
    "            .rename({sample_col: 'n_samples_effective'}, axis=1))\n",
    "    df = df.merge(temp,\n",
    "                  how='left',\n",
    "                  on='transcript_id')\n",
    "    df['perc_effective_isoforms'] = (df['n_samples_effective']/n_samples)*100\n",
    "\n",
    "    # fill nans w/ 0\n",
    "    df['perc_effective_isoforms'] = df['perc_effective_isoforms'].fillna(0, inplace=True)\n",
    "    \n",
    "    # information obout expression across samples\n",
    "    df['n_exp_samples'] = (df.groupby('transcript_id')[sample_col]\n",
    "                       .transform('nunique'))\n",
    "    df['pi_std'] = (df.groupby('transcript_id')['pi']\n",
    "                       .transform(lambda x: x.std(ddof=1, skipna=True)))\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "id": "a39bf58b-c986-41c3-b1e8-e9bed62d964a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/lib/python3.7/site-packages/pandas/core/arraylike.py:364: RuntimeWarning: divide by zero encountered in log2\n",
      "  result = getattr(ufunc, method)(*inputs, **kwargs)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>perplexity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>brain</td>\n",
       "      <td>A_1</td>\n",
       "      <td>2.582907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>brain</td>\n",
       "      <td>A_2</td>\n",
       "      <td>2.582907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>brain</td>\n",
       "      <td>A_3</td>\n",
       "      <td>2.582907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>brain</td>\n",
       "      <td>A_4</td>\n",
       "      <td>2.582907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>brain</td>\n",
       "      <td>A_5</td>\n",
       "      <td>2.582907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>brain</td>\n",
       "      <td>A_6</td>\n",
       "      <td>2.582907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>brain</td>\n",
       "      <td>A_7</td>\n",
       "      <td>2.582907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>brain</td>\n",
       "      <td>A_8</td>\n",
       "      <td>2.582907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>heart</td>\n",
       "      <td>A_1</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>heart</td>\n",
       "      <td>A_2</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>heart</td>\n",
       "      <td>A_3</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>heart</td>\n",
       "      <td>A_4</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>heart</td>\n",
       "      <td>A_5</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>heart</td>\n",
       "      <td>A_6</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>heart</td>\n",
       "      <td>A_7</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>heart</td>\n",
       "      <td>A_8</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>kidney</td>\n",
       "      <td>A_1</td>\n",
       "      <td>5.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>kidney</td>\n",
       "      <td>A_2</td>\n",
       "      <td>5.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>kidney</td>\n",
       "      <td>A_3</td>\n",
       "      <td>5.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>kidney</td>\n",
       "      <td>A_4</td>\n",
       "      <td>5.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>kidney</td>\n",
       "      <td>A_5</td>\n",
       "      <td>5.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>kidney</td>\n",
       "      <td>A_6</td>\n",
       "      <td>5.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>kidney</td>\n",
       "      <td>A_7</td>\n",
       "      <td>5.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>kidney</td>\n",
       "      <td>A_8</td>\n",
       "      <td>5.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>lungs</td>\n",
       "      <td>A_1</td>\n",
       "      <td>4.377939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>lungs</td>\n",
       "      <td>A_2</td>\n",
       "      <td>4.377939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>lungs</td>\n",
       "      <td>A_3</td>\n",
       "      <td>4.377939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>lungs</td>\n",
       "      <td>A_4</td>\n",
       "      <td>4.377939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>lungs</td>\n",
       "      <td>A_5</td>\n",
       "      <td>4.377939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>lungs</td>\n",
       "      <td>A_6</td>\n",
       "      <td>4.377939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>lungs</td>\n",
       "      <td>A_7</td>\n",
       "      <td>4.377939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>lungs</td>\n",
       "      <td>A_8</td>\n",
       "      <td>4.377939</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample transcript_id  perplexity\n",
       "8    brain           A_1    2.582907\n",
       "9    brain           A_2    2.582907\n",
       "10   brain           A_3    2.582907\n",
       "11   brain           A_4    2.582907\n",
       "12   brain           A_5    2.582907\n",
       "13   brain           A_6    2.582907\n",
       "14   brain           A_7    2.582907\n",
       "15   brain           A_8    2.582907\n",
       "0    heart           A_1    2.000000\n",
       "1    heart           A_2    2.000000\n",
       "2    heart           A_3    2.000000\n",
       "3    heart           A_4    2.000000\n",
       "4    heart           A_5    2.000000\n",
       "5    heart           A_6    2.000000\n",
       "6    heart           A_7    2.000000\n",
       "7    heart           A_8    2.000000\n",
       "24  kidney           A_1    5.000000\n",
       "25  kidney           A_2    5.000000\n",
       "26  kidney           A_3    5.000000\n",
       "27  kidney           A_4    5.000000\n",
       "28  kidney           A_5    5.000000\n",
       "29  kidney           A_6    5.000000\n",
       "30  kidney           A_7    5.000000\n",
       "31  kidney           A_8    5.000000\n",
       "16   lungs           A_1    4.377939\n",
       "17   lungs           A_2    4.377939\n",
       "18   lungs           A_3    4.377939\n",
       "19   lungs           A_4    4.377939\n",
       "20   lungs           A_5    4.377939\n",
       "21   lungs           A_6    4.377939\n",
       "22   lungs           A_7    4.377939\n",
       "23   lungs           A_8    4.377939"
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = dummy_data_2()\n",
    "sample_col = 'sample'\n",
    "df = workflow_2(df, sample_col)\n",
    "\n",
    "df[['transcript_id', 'pi_std']].drop_duplicates()\n",
    "\n",
    "df[['sample', 'transcript_id', 'perplexity']].sort_values(by=['sample', 'transcript_id']).drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "f1c85f39-7c64-464e-9e31-477ec8af0422",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>pi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A_7</td>\n",
       "      <td>0.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>A_7</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A_7</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A_7</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   transcript_id    pi\n",
       "14           A_7  0.45\n",
       "30           A_7  0.00\n",
       "6            A_7  0.00\n",
       "22           A_7  0.00"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df[['transcript_id', 'pi']].sort_values(by='transcript_id')\n",
    "temp.loc[temp.transcript_id == 'A_7']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "f48f575b-aa13-46f3-acb7-65231c76ba0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.16007810593582122\n",
      "0.09574271077563382\n",
      "0.09451631252505217\n",
      "0.21\n",
      "0.25\n",
      "0.09521904571390467\n",
      "0.22500000000000003\n",
      "0.03\n"
     ]
    }
   ],
   "source": [
    "print(np.std([0.5, 0.45, 0.2, 0.2], ddof=1))\n",
    "print(np.std([0, 0, 0.1, 0.2], ddof=1))\n",
    "print(np.std([0, 0, 0.08, 0.2], ddof=1))\n",
    "print(np.std([0.5, 0.1, 0.02, 0.2], ddof=1))\n",
    "print(np.std([0, 0, 0.5, 0], ddof=1))\n",
    "print(np.std([0, 0, 0.04, 0.2], ddof=1))\n",
    "print(np.std([0, 0.45, 0, 0], ddof=1))\n",
    "print(np.std([0, 0, 0.06, 0], ddof=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "8e6ae7fb-09eb-4427-9a87-f3ebf31bcfae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.16007810593582122\n",
      "0.09574271077563382\n",
      "0.09451631252505217\n",
      "0.21\n"
     ]
    }
   ],
   "source": [
    "for arr in [[0.5, 0.45, 0.2, 0.2],\n",
    "            [0,0,0.1,0.2],\n",
    "            [0,0,0.08,0.2],\n",
    "            [0.5,0.1,0.02,0.2]]:\n",
    "    print(np.std(arr, ddof=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a1127c9-063d-422b-a6f6-bacf614b9c72",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_pi(df):\n",
    "    \"\"\"\n",
    "    generate pi values (isoform ratios) from raw counts\n",
    "    \"\"\"\n",
    "    df['gene_counts'] = df.groupby('gene_id').transform('sum')['counts']\n",
    "    df['pi'] = df['counts']/df['gene_counts']\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "4a87bfe1-8541-4f45-865d-8956e8ed45ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_dummy_data_counts():\n",
    "    df = pd.DataFrame()\n",
    "    df['gene_id'] = ['A' for i in range(2)]+\\\n",
    "                    ['B' for i in range(7)]+\\\n",
    "                    ['C' for i in range(8)]\n",
    "    df['transcript_id'] = [f'A_{i+1}' for i in range(2)]+\\\n",
    "                          [f'B_{i+1}' for i in range(7)]+\\\n",
    "                          [f'C_{i+1}' for i in range(8)]\n",
    "    df['counts'] = ([47, 47]+ \n",
    "                    [i*28 for i in [0.4, 0.35, 0.10, 0.07, 0.04, 0.03, 0.01]]+\n",
    "                    [0.125 for i in range(8)])\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "id": "3e5c57ca-7b8f-4972-9951-0f081bc631e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>counts</th>\n",
       "      <th>gene_counts</th>\n",
       "      <th>pi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>47.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>47.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B</td>\n",
       "      <td>B_1</td>\n",
       "      <td>11.2</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B</td>\n",
       "      <td>B_2</td>\n",
       "      <td>9.8</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B</td>\n",
       "      <td>B_3</td>\n",
       "      <td>2.8</td>\n",
       "      <td>28.0</td>\n",
       "      <td>0.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id transcript_id  counts gene_counts    pi\n",
       "0       A           A_1    47.0        94.0   0.5\n",
       "1       A           A_2    47.0        94.0   0.5\n",
       "2       B           B_1    11.2        28.0   0.4\n",
       "3       B           B_2     9.8        28.0  0.35\n",
       "4       B           B_3     2.8        28.0   0.1"
      ]
     },
     "execution_count": 257,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = get_dummy_data_counts()\n",
    "df['gene_counts'] = df.groupby('gene_id').transform('sum')['counts']\n",
    "df['pi'] = df['counts']/df['gene_counts']\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "id": "5c5ceac6-3f2a-4cbe-8dfc-641e83728255",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[47,\n",
       " 47,\n",
       " 11.200000000000001,\n",
       " 9.799999999999999,\n",
       " 2.8000000000000003,\n",
       " 1.9600000000000002,\n",
       " 1.12,\n",
       " 0.84,\n",
       " 0.28,\n",
       " 0.125,\n",
       " 0.125,\n",
       " 0.125,\n",
       " 0.125,\n",
       " 0.125,\n",
       " 0.125,\n",
       " 0.125,\n",
       " 0.125]"
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l = ([47, 47]+ \n",
    "           [i*28 for i in [0.4, 0.35, 0.10, 0.07, 0.04, 0.03, 0.01]]+\n",
    "             [0.125 for i in range(8)])\n",
    "# flatten_list(l)\n",
    "l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08ce640d-2661-4544-93b9-bd27ce5e964c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "226a2173-740f-4ff7-b16c-20dbfcb5926f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
