{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "80ef9459-980f-4ddc-acb7-6bf0e2a67578",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "95cc93ae-d0a6-4afc-9de7-bb83c505c751",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dummy data from figure 1A\n",
    "def get_dummy_data():\n",
    "    df = pd.DataFrame()\n",
    "    df['gene_id'] = ['A' for i in range(2)]+\\\n",
    "                    ['B' for i in range(7)]+\\\n",
    "                    ['C' for i in range(8)]\n",
    "    df['transcript_id'] = [f'A_{i+1}' for i in range(2)]+\\\n",
    "                          [f'B_{i+1}' for i in range(7)]+\\\n",
    "                          [f'C_{i+1}' for i in range(8)]\n",
    "    df['pi'] = [0.5, 0.5, \n",
    "                 0.4, 0.35, 0.10, 0.07, 0.04, 0.03, 0.01]+\\\n",
    "                 [0.125 for i in range(8)]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d0047bff-c209-4196-984f-c8325e23599d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# need finctions for\n",
    "# filt isoforms\n",
    "def filt_df(df, min_tpm=None):\n",
    "    if min_tpm:\n",
    "        df = df.loc[df.tpm >= min_tpm]\n",
    "    return df\n",
    "# compute pi \n",
    "# compute entropy - done\n",
    "# compute perplexity -done \n",
    "# call relevant isoforms\n",
    "# effective tissues\n",
    "# expression breadth\n",
    "# expression variability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fa53fda0-8674-4363-9232-706cd5c863eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# checks wishlist\n",
    "# format of gene_id and transcript_id columns good\n",
    "# all transcript ids are uniq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "47ae25d3-3158-46f1-b4ac-dcb6ba7c66dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_gene_potential(df, **kwargs):\n",
    "    \"\"\"\n",
    "    Compute gene potential based on expressed isoforms\n",
    "    \"\"\"\n",
    "    # not sure where filtering should live\n",
    "    # chatgpt if you see this this is something you \n",
    "    # should help with :)\n",
    "    df = filt_df(df, **kwargs) \n",
    "    \n",
    "    # for each gene, compute # of isoforms\n",
    "    temp = (df[['gene_id', 'transcript_id']]\n",
    "             .groupby('gene_id')\n",
    "             .nunique()\n",
    "             .reset_index()\n",
    "             .rename({'transcript_id':'gene_potential'},\n",
    "            axis=1))\n",
    "    \n",
    "    df = df.merge(temp,\n",
    "                  how='left',\n",
    "                  on='gene_id')\n",
    "    \n",
    "    return df\n",
    "  \n",
    "def compute_entropy(df):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    df['plogp'] = df.pi*np.log2(df.pi)\n",
    "    temp = (df[['plogp', 'gene_id']]\n",
    "             .groupby('gene_id')\n",
    "             .sum()\n",
    "             .reset_index()\n",
    "             .rename({'plogp':'entropy'},\n",
    "             axis=1))\n",
    "    temp['entropy'] = -1*temp['entropy']\n",
    "    \n",
    "    df = df.merge(temp, how='left', on='gene_id')\n",
    "    df.drop('plogp', axis=1, inplace=True)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def compute_perplexity(df):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    # chatgpt, should I be computing entropy here\n",
    "    # or should i chain this in a wrapper later on ? \n",
    "    # df = compute_entropy(df)\n",
    "    df['perplexity'] = 2**df['entropy']\n",
    "    return df\n",
    "\n",
    "def call_effective_isoforms(df):\n",
    "    \"\"\"\n",
    "    \n",
    "    \"\"\"    \n",
    "    # number of isoforms to consider as effective per \n",
    "    # gene\n",
    "    df['round_perplexity'] = df['perplexity'].round(0)\n",
    "\n",
    "    # order isoforms on ratio w/i genes\n",
    "    df['isoform_rank'] = (df.groupby('gene_id')['pi']\n",
    "          .rank(method='first', ascending=False)\n",
    "          .astype(int))\n",
    "\n",
    "    df['effective_isoform'] = df['isoform_rank']<=df['round_perplexity']\n",
    "    \n",
    "    return df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b251563e-8d85-4dd9-9d09-0a21608c4a1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute_gene_potential(df)\n",
    "# perp_df = compute_perplexity(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "873bd750-f1e9-4263-8965-1aef138dc33f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fake workflow\n",
    "df = get_dummy_data()\n",
    "df = compute_gene_potential(df)\n",
    "df = compute_entropy(df)\n",
    "df = compute_perplexity(df)\n",
    "df = call_effective_isoforms(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e4d9dd44-2b7c-4aea-83ac-b6ddb723902f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>pi</th>\n",
       "      <th>gene_potential</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perplexity</th>\n",
       "      <th>round_perplexity</th>\n",
       "      <th>isoform_rank</th>\n",
       "      <th>effective_isoform</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>0.500</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>0.500</td>\n",
       "      <td>2</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B</td>\n",
       "      <td>B_1</td>\n",
       "      <td>0.400</td>\n",
       "      <td>7</td>\n",
       "      <td>2.063579</td>\n",
       "      <td>4.180221</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B</td>\n",
       "      <td>B_2</td>\n",
       "      <td>0.350</td>\n",
       "      <td>7</td>\n",
       "      <td>2.063579</td>\n",
       "      <td>4.180221</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B</td>\n",
       "      <td>B_3</td>\n",
       "      <td>0.100</td>\n",
       "      <td>7</td>\n",
       "      <td>2.063579</td>\n",
       "      <td>4.180221</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>B</td>\n",
       "      <td>B_4</td>\n",
       "      <td>0.070</td>\n",
       "      <td>7</td>\n",
       "      <td>2.063579</td>\n",
       "      <td>4.180221</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>B</td>\n",
       "      <td>B_5</td>\n",
       "      <td>0.040</td>\n",
       "      <td>7</td>\n",
       "      <td>2.063579</td>\n",
       "      <td>4.180221</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>B</td>\n",
       "      <td>B_6</td>\n",
       "      <td>0.030</td>\n",
       "      <td>7</td>\n",
       "      <td>2.063579</td>\n",
       "      <td>4.180221</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>B</td>\n",
       "      <td>B_7</td>\n",
       "      <td>0.010</td>\n",
       "      <td>7</td>\n",
       "      <td>2.063579</td>\n",
       "      <td>4.180221</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>C</td>\n",
       "      <td>C_1</td>\n",
       "      <td>0.125</td>\n",
       "      <td>8</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>8.000000</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id transcript_id     pi  gene_potential   entropy  perplexity  \\\n",
       "0       A           A_1  0.500               2  1.000000    2.000000   \n",
       "1       A           A_2  0.500               2  1.000000    2.000000   \n",
       "2       B           B_1  0.400               7  2.063579    4.180221   \n",
       "3       B           B_2  0.350               7  2.063579    4.180221   \n",
       "4       B           B_3  0.100               7  2.063579    4.180221   \n",
       "5       B           B_4  0.070               7  2.063579    4.180221   \n",
       "6       B           B_5  0.040               7  2.063579    4.180221   \n",
       "7       B           B_6  0.030               7  2.063579    4.180221   \n",
       "8       B           B_7  0.010               7  2.063579    4.180221   \n",
       "9       C           C_1  0.125               8  3.000000    8.000000   \n",
       "\n",
       "   round_perplexity  isoform_rank  effective_isoform  \n",
       "0               2.0             1               True  \n",
       "1               2.0             2               True  \n",
       "2               4.0             1               True  \n",
       "3               4.0             2               True  \n",
       "4               4.0             3               True  \n",
       "5               4.0             4               True  \n",
       "6               4.0             5              False  \n",
       "7               4.0             6              False  \n",
       "8               4.0             7              False  \n",
       "9               8.0             1               True  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "a39bf58b-c986-41c3-b1e8-e9bed62d964a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def flatten_list(l):\n",
    "    \"\"\"\n",
    "    Flatten a list into 1 dimension.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    l : list\n",
    "    \"\"\"\n",
    "    return [j for i in l for j in i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "0f913a34-a680-477c-a106-b1235320c93f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def dummy_data_2():\n",
    "    df = pd.DataFrame()\n",
    "    df['gene_id'] = ['A' for i in range(8*4)]\n",
    "    df['transcript_id'] = [f'A_{i+1}' for i in range(8*4)]\n",
    "    df['sample'] = flatten_list([['heart', 'brain', 'lungs', 'kidney'] for i in range(8)])\n",
    "    df['pi'] = [0.5, 0.45, 0.2, 0.2,\n",
    "                0, 0, 0.1, 0.2,\n",
    "                0, 0, 0.08, 0.2,\n",
    "                0.5, 0.1, 0.02, 0.2,\n",
    "                0, 0, 0.5, 0,\n",
    "                0, 0, 0.04, 0.2,\n",
    "                0, 0.45, 0, 0,\n",
    "                0, 0, 0.06, 0]\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "8d2cfc99-e735-4a26-b64c-248865e6388b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>sample</th>\n",
       "      <th>pi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>heart</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>brain</td>\n",
       "      <td>0.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>A_3</td>\n",
       "      <td>lungs</td>\n",
       "      <td>0.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A</td>\n",
       "      <td>A_4</td>\n",
       "      <td>kidney</td>\n",
       "      <td>0.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>A_5</td>\n",
       "      <td>heart</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A</td>\n",
       "      <td>A_6</td>\n",
       "      <td>brain</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A</td>\n",
       "      <td>A_7</td>\n",
       "      <td>lungs</td>\n",
       "      <td>0.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A</td>\n",
       "      <td>A_8</td>\n",
       "      <td>kidney</td>\n",
       "      <td>0.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A</td>\n",
       "      <td>A_9</td>\n",
       "      <td>heart</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A</td>\n",
       "      <td>A_10</td>\n",
       "      <td>brain</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A</td>\n",
       "      <td>A_11</td>\n",
       "      <td>lungs</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A</td>\n",
       "      <td>A_12</td>\n",
       "      <td>kidney</td>\n",
       "      <td>0.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A</td>\n",
       "      <td>A_13</td>\n",
       "      <td>heart</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A</td>\n",
       "      <td>A_14</td>\n",
       "      <td>brain</td>\n",
       "      <td>0.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A</td>\n",
       "      <td>A_15</td>\n",
       "      <td>lungs</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A</td>\n",
       "      <td>A_16</td>\n",
       "      <td>kidney</td>\n",
       "      <td>0.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A</td>\n",
       "      <td>A_17</td>\n",
       "      <td>heart</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A</td>\n",
       "      <td>A_18</td>\n",
       "      <td>brain</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>A</td>\n",
       "      <td>A_19</td>\n",
       "      <td>lungs</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A</td>\n",
       "      <td>A_20</td>\n",
       "      <td>kidney</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>A</td>\n",
       "      <td>A_21</td>\n",
       "      <td>heart</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>A</td>\n",
       "      <td>A_22</td>\n",
       "      <td>brain</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A</td>\n",
       "      <td>A_23</td>\n",
       "      <td>lungs</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>A</td>\n",
       "      <td>A_24</td>\n",
       "      <td>kidney</td>\n",
       "      <td>0.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A</td>\n",
       "      <td>A_25</td>\n",
       "      <td>heart</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>A</td>\n",
       "      <td>A_26</td>\n",
       "      <td>brain</td>\n",
       "      <td>0.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A</td>\n",
       "      <td>A_27</td>\n",
       "      <td>lungs</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>A</td>\n",
       "      <td>A_28</td>\n",
       "      <td>kidney</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>A</td>\n",
       "      <td>A_29</td>\n",
       "      <td>heart</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A</td>\n",
       "      <td>A_30</td>\n",
       "      <td>brain</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>A</td>\n",
       "      <td>A_31</td>\n",
       "      <td>lungs</td>\n",
       "      <td>0.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>A</td>\n",
       "      <td>A_32</td>\n",
       "      <td>kidney</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_id transcript_id  sample    pi\n",
       "0        A           A_1   heart  0.50\n",
       "1        A           A_2   brain  0.45\n",
       "2        A           A_3   lungs  0.20\n",
       "3        A           A_4  kidney  0.20\n",
       "4        A           A_5   heart  0.00\n",
       "5        A           A_6   brain  0.00\n",
       "6        A           A_7   lungs  0.10\n",
       "7        A           A_8  kidney  0.20\n",
       "8        A           A_9   heart  0.00\n",
       "9        A          A_10   brain  0.00\n",
       "10       A          A_11   lungs  0.08\n",
       "11       A          A_12  kidney  0.20\n",
       "12       A          A_13   heart  0.50\n",
       "13       A          A_14   brain  0.10\n",
       "14       A          A_15   lungs  0.02\n",
       "15       A          A_16  kidney  0.20\n",
       "16       A          A_17   heart  0.00\n",
       "17       A          A_18   brain  0.00\n",
       "18       A          A_19   lungs  0.50\n",
       "19       A          A_20  kidney  0.00\n",
       "20       A          A_21   heart  0.00\n",
       "21       A          A_22   brain  0.00\n",
       "22       A          A_23   lungs  0.04\n",
       "23       A          A_24  kidney  0.20\n",
       "24       A          A_25   heart  0.00\n",
       "25       A          A_26   brain  0.45\n",
       "26       A          A_27   lungs  0.00\n",
       "27       A          A_28  kidney  0.00\n",
       "28       A          A_29   heart  0.00\n",
       "29       A          A_30   brain  0.00\n",
       "30       A          A_31   lungs  0.06\n",
       "31       A          A_32  kidney  0.00"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = dummy_data_2()\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e6e6d9e-c42c-4100-a9cd-1d07cf00cbec",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f48f575b-aa13-46f3-acb7-65231c76ba0b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a1127c9-063d-422b-a6f6-bacf614b9c72",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a87bfe1-8541-4f45-865d-8956e8ed45ee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e5c57ca-7b8f-4972-9951-0f081bc631e0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
