{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b6d13a62-75d1-49fb-a05c-dd68e009cebd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils import * "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "439a7673-7601-4c37-b47d-d2729e5709bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_dummy_data_counts():\n",
    "    df = pd.DataFrame()\n",
    "    df['gene_id'] = ['A' for i in range(2)]+\\\n",
    "                    ['B' for i in range(7)]+\\\n",
    "                    ['C' for i in range(8)]\n",
    "    df['transcript_id'] = [f'A_{i+1}' for i in range(2)]+\\\n",
    "                          [f'B_{i+1}' for i in range(7)]+\\\n",
    "                          [f'C_{i+1}' for i in range(8)]\n",
    "    df['counts'] = ([47, 47]+ \n",
    "                    [i*28 for i in [0.4, 0.35, 0.10, 0.07, 0.04, 0.03, 0.01]]+\n",
    "                    [0.125 for i in range(8)])\n",
    "        \n",
    "    return df\n",
    "\n",
    "def dummy_counts_data_sample():\n",
    "    \"\"\"\n",
    "    Generate a dummy multi-sample dataset with counts instead of pi values.\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        Columns: 'gene_id', 'transcript_id', 'sample', 'counts'\n",
    "    \"\"\"\n",
    "    import pandas as pd\n",
    "\n",
    "    genes = ['A']  # single gene for simplicity\n",
    "    transcripts = [f'A_{i+1}' for i in range(8)]\n",
    "    samples = ['heart', 'brain', 'lungs', 'kidney']\n",
    "    \n",
    "    # create repeated combinations\n",
    "    df = pd.DataFrame({\n",
    "        'gene_id': flatten_list([[g]*len(samples) for g in genes*len(transcripts)]),\n",
    "        'transcript_id': flatten_list([[t]*len(samples) for t in transcripts]),\n",
    "        'sample': samples * len(transcripts),\n",
    "    })\n",
    "    \n",
    "    # assign dummy counts\n",
    "    df['counts'] = [\n",
    "        50, 45, 20, 20,   # A_1\n",
    "        0, 0, 10, 20,     # A_2\n",
    "        0, 0, 8, 20,      # A_3\n",
    "        50, 10, 2, 20,    # A_4\n",
    "        0, 0, 50, 0,      # A_5\n",
    "        0, 0, 4, 20,      # A_6\n",
    "        0, 45, 0, 0,      # A_7\n",
    "        0, 0, 6, 0        # A_8\n",
    "    ]\n",
    "    \n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "630121b6-afcf-4b44-88a6-132ac8b43345",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # all positive counts\n",
    "# df = get_dummy_data_counts()\n",
    "# validate_counts_input(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d7bee416-b97c-4e30-8d58-1fd5620da159",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # w/ neg counts\n",
    "# df = get_dummy_data_counts()\n",
    "# df['counts'] = df['counts'].tolist()[:-1]+[-1]\n",
    "# validate_counts_input(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2b18dd53-cbe8-4991-97c3-6455b4bf477d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # w/o sample col when it needs one\n",
    "# df = get_dummy_data_counts()\n",
    "# validate_counts_input(df, sample_col='sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d10563ca-1497-4244-bcdd-2495fc998a57",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # w/o gene id col when it needs one\n",
    "# df = get_dummy_data_counts()\n",
    "# validate_counts_input(df, sample_col='sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "04334334-2688-4cb1-ab03-0ace089b4556",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # dummy data from figure 1A\n",
    "# def get_dummy_data():\n",
    "#     df = pd.DataFrame()\n",
    "#     df['gene_id'] = ['A' for i in range(2)]+\\\n",
    "#                     ['B' for i in range(7)]+\\\n",
    "#                     ['C' for i in range(8)]\n",
    "#     df['transcript_id'] = [f'A_{i+1}' for i in range(2)]+\\\n",
    "#                           [f'B_{i+1}' for i in range(7)]+\\\n",
    "#                           [f'C_{i+1}' for i in range(8)]\n",
    "#     df['pi'] = [0.5, 0.5, \n",
    "#                  0.4, 0.35, 0.10, 0.07, 0.04, 0.03, 0.01]+\\\n",
    "#                  [0.125 for i in range(8)]\n",
    "#     return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7982627e-e5e7-4041-abc1-b9224ba31647",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  gene_id  gene_potential   entropy  perplexity  n_effective_isoforms\n",
      "0       A               2  1.000000    2.000000                   2.0\n",
      "2       B               7  2.063579    4.180221                   4.0\n",
      "9       C               8  3.000000    8.000000                   8.0\n",
      "\n",
      "   gene_id transcript_id     pi  effective_isoform\n",
      "0        A           A_1  0.500               True\n",
      "1        A           A_2  0.500               True\n",
      "2        B           B_1  0.400               True\n",
      "3        B           B_2  0.350               True\n",
      "4        B           B_3  0.100               True\n",
      "5        B           B_4  0.070               True\n",
      "6        B           B_5  0.040              False\n",
      "7        B           B_6  0.030              False\n",
      "8        B           B_7  0.010              False\n",
      "9        C           C_1  0.125               True\n",
      "10       C           C_2  0.125               True\n",
      "11       C           C_3  0.125               True\n",
      "12       C           C_4  0.125               True\n",
      "13       C           C_5  0.125               True\n",
      "14       C           C_6  0.125               True\n",
      "15       C           C_7  0.125               True\n",
      "16       C           C_8  0.125               True\n"
     ]
    }
   ],
   "source": [
    "df = get_dummy_data_counts()\n",
    "df = compute_global_isoform_metrics(df)\n",
    "print(df[['gene_id', 'gene_potential', 'entropy', 'perplexity', 'n_effective_isoforms']].drop_duplicates())\n",
    "print()\n",
    "print(df[['gene_id', 'transcript_id', 'pi', 'effective_isoform']])\n",
    "df.to_csv('example_bulk_metrics.tsv', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5da7c52c-47b7-4dc7-b34e-2896b848ebde",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30246531"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "30246531"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "84676fda-0065-455b-9cd9-cda016bb3f59",
   "metadata": {},
   "outputs": [],
   "source": [
    "def collapse_counts_by_feature(df,\n",
    "                               feature_col=TRANSCRIPT_COL,\n",
    "                               gene_col=GENE_COL,\n",
    "                               sample_col=None):\n",
    "    \"\"\"\n",
    "    Collapse counts by a feature (e.g. ORF, transcript) instead of transcript.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        Input table with counts at transcript level.\n",
    "    feature_col : str\n",
    "        Alternative feature column to collapse to (e.g. 'orf_id').\n",
    "    gene_col : str\n",
    "        Column identifying genes.\n",
    "    sample_col : str, optional\n",
    "        Sample column. If None, assumes single-sample bulk.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        Collapsed df with summed counts per feature.\n",
    "    \"\"\"\n",
    "    group_cols = [gene_col, feature_col]\n",
    "    if sample_col is not None:\n",
    "        group_cols.append(sample_col)\n",
    "        \n",
    "    import pdb; pdb.set_trace()\n",
    "        \n",
    "    # Sum counts for all transcripts mapping to the same feature\n",
    "    out = (df.groupby(group_cols, as_index=False)['counts']\n",
    "          .sum())\n",
    "\n",
    "    return out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fa309802-da0a-42af-8f98-61173956b56d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   gene_id  sample  gene_potential   entropy  perplexity  n_effective_isoforms\n",
      "0        A   heart               8  1.000000    2.000000                   2.0\n",
      "8        A   brain               8  1.368996    2.582907                   3.0\n",
      "16       A   lungs               8  2.130252    4.377939                   4.0\n",
      "24       A  kidney               8  2.321928    5.000000                   5.0\n",
      "\n",
      "  gene_id transcript_id  expression_breadth  expression_var\n",
      "0       A           A_1               100.0        0.160078\n",
      "1       A           A_2                50.0        0.095743\n",
      "2       A           A_3                50.0        0.094516\n",
      "3       A           A_4                75.0        0.210000\n",
      "4       A           A_5                25.0        0.250000\n",
      "5       A           A_6                25.0        0.095219\n",
      "6       A           A_7                25.0        0.225000\n",
      "7       A           A_8                 0.0        0.030000\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>sample</th>\n",
       "      <th>counts</th>\n",
       "      <th>pi</th>\n",
       "      <th>gene_potential</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perplexity</th>\n",
       "      <th>n_effective_isoforms</th>\n",
       "      <th>isoform_rank</th>\n",
       "      <th>effective_isoform</th>\n",
       "      <th>n_samples_effective</th>\n",
       "      <th>expression_breadth</th>\n",
       "      <th>n_exp_samples</th>\n",
       "      <th>expression_var</th>\n",
       "      <th>avg_transcript_id_counts</th>\n",
       "      <th>avg_gene_id_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>heart</td>\n",
       "      <td>50</td>\n",
       "      <td>0.5</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.160078</td>\n",
       "      <td>33.75</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.095743</td>\n",
       "      <td>15.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>A_3</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.094516</td>\n",
       "      <td>14.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A</td>\n",
       "      <td>A_4</td>\n",
       "      <td>heart</td>\n",
       "      <td>50</td>\n",
       "      <td>0.5</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.210000</td>\n",
       "      <td>20.50</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>A_5</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>50.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id transcript_id sample  counts   pi  gene_potential  entropy  \\\n",
       "0       A           A_1  heart      50  0.5               8      1.0   \n",
       "1       A           A_2  heart       0  0.0               8      1.0   \n",
       "2       A           A_3  heart       0  0.0               8      1.0   \n",
       "3       A           A_4  heart      50  0.5               8      1.0   \n",
       "4       A           A_5  heart       0  0.0               8      1.0   \n",
       "\n",
       "   perplexity  n_effective_isoforms  isoform_rank  effective_isoform  \\\n",
       "0         2.0                   2.0             1               True   \n",
       "1         2.0                   2.0             3              False   \n",
       "2         2.0                   2.0             4              False   \n",
       "3         2.0                   2.0             2               True   \n",
       "4         2.0                   2.0             5              False   \n",
       "\n",
       "   n_samples_effective  expression_breadth  n_exp_samples  expression_var  \\\n",
       "0                  4.0               100.0              4        0.160078   \n",
       "1                  2.0                50.0              4        0.095743   \n",
       "2                  2.0                50.0              4        0.094516   \n",
       "3                  3.0                75.0              4        0.210000   \n",
       "4                  1.0                25.0              4        0.250000   \n",
       "\n",
       "   avg_transcript_id_counts  avg_gene_id_counts  \n",
       "0                     33.75               100.0  \n",
       "1                     15.00               100.0  \n",
       "2                     14.00               100.0  \n",
       "3                     20.50               100.0  \n",
       "4                     50.00               100.0  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = dummy_counts_data_sample()\n",
    "df = compute_multi_sample_isoform_metrics(df, 'sample')\n",
    "print(df[['gene_id', 'sample', 'gene_potential', 'entropy', 'perplexity', 'n_effective_isoforms']].drop_duplicates())\n",
    "print()\n",
    "print(df[['gene_id', 'transcript_id','expression_breadth', 'expression_var']].drop_duplicates())\n",
    "df.to_csv('example_sample_metrics.tsv', sep='\\t')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "910769bf-ba44-44ca-af39-0f03a24426fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33.75"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(50+45+20+20)/4\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fb470f44-6475-4ee2-a110-6f7375e0f155",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>sample</th>\n",
       "      <th>counts</th>\n",
       "      <th>pi</th>\n",
       "      <th>gene_potential</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perplexity</th>\n",
       "      <th>n_effective_isoforms</th>\n",
       "      <th>isoform_rank</th>\n",
       "      <th>effective_isoform</th>\n",
       "      <th>n_samples_effective</th>\n",
       "      <th>expression_breadth</th>\n",
       "      <th>n_exp_samples</th>\n",
       "      <th>expression_var</th>\n",
       "      <th>avg_transcript_id_counts</th>\n",
       "      <th>avg_gene_id_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>heart</td>\n",
       "      <td>50</td>\n",
       "      <td>0.50</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.160078</td>\n",
       "      <td>33.75</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>brain</td>\n",
       "      <td>45</td>\n",
       "      <td>0.45</td>\n",
       "      <td>8</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.160078</td>\n",
       "      <td>33.75</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>lungs</td>\n",
       "      <td>20</td>\n",
       "      <td>0.20</td>\n",
       "      <td>8</td>\n",
       "      <td>2.130252</td>\n",
       "      <td>4.377939</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.160078</td>\n",
       "      <td>33.75</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>kidney</td>\n",
       "      <td>20</td>\n",
       "      <td>0.20</td>\n",
       "      <td>8</td>\n",
       "      <td>2.321928</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.160078</td>\n",
       "      <td>33.75</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_id transcript_id  sample  counts    pi  gene_potential   entropy  \\\n",
       "0        A           A_1   heart      50  0.50               8  1.000000   \n",
       "8        A           A_1   brain      45  0.45               8  1.368996   \n",
       "16       A           A_1   lungs      20  0.20               8  2.130252   \n",
       "24       A           A_1  kidney      20  0.20               8  2.321928   \n",
       "\n",
       "    perplexity  n_effective_isoforms  isoform_rank  effective_isoform  \\\n",
       "0     2.000000                   2.0             1               True   \n",
       "8     2.582907                   3.0             1               True   \n",
       "16    4.377939                   4.0             2               True   \n",
       "24    5.000000                   5.0             1               True   \n",
       "\n",
       "    n_samples_effective  expression_breadth  n_exp_samples  expression_var  \\\n",
       "0                   4.0               100.0              4        0.160078   \n",
       "8                   4.0               100.0              4        0.160078   \n",
       "16                  4.0               100.0              4        0.160078   \n",
       "24                  4.0               100.0              4        0.160078   \n",
       "\n",
       "    avg_transcript_id_counts  avg_gene_id_counts  \n",
       "0                      33.75               100.0  \n",
       "8                      33.75               100.0  \n",
       "16                     33.75               100.0  \n",
       "24                     33.75               100.0  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.transcript_id=='A_1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f49dc862-4b42-4b86-8d76-d6a9e11956b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>perplexity</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>heart</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A</td>\n",
       "      <td>4.377939</td>\n",
       "      <td>lungs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>kidney</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_id  perplexity  sample\n",
       "0        A    2.000000   heart\n",
       "8        A    2.582907   brain\n",
       "16       A    4.377939   lungs\n",
       "24       A    5.000000  kidney"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['gene_id', 'perplexity', 'sample']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5618e9c8-eb4f-4b76-9254-21fa85f35ffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df[['gene_id', 'gene_counts', 'sample']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "84eaf2cd-89f4-4391-8cad-eac41d030d8d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>sample</th>\n",
       "      <th>counts</th>\n",
       "      <th>pi</th>\n",
       "      <th>gene_potential</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perplexity</th>\n",
       "      <th>n_effective_isoforms</th>\n",
       "      <th>isoform_rank</th>\n",
       "      <th>effective_isoform</th>\n",
       "      <th>n_samples_effective</th>\n",
       "      <th>expression_breadth</th>\n",
       "      <th>n_exp_samples</th>\n",
       "      <th>expression_var</th>\n",
       "      <th>avg_transcript_id_counts</th>\n",
       "      <th>avg_gene_id_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>heart</td>\n",
       "      <td>50</td>\n",
       "      <td>0.5</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.160078</td>\n",
       "      <td>33.75</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.095743</td>\n",
       "      <td>15.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>A_3</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.094516</td>\n",
       "      <td>14.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A</td>\n",
       "      <td>A_4</td>\n",
       "      <td>heart</td>\n",
       "      <td>50</td>\n",
       "      <td>0.5</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.210000</td>\n",
       "      <td>20.50</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>A_5</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>50.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A</td>\n",
       "      <td>A_6</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.095219</td>\n",
       "      <td>12.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A</td>\n",
       "      <td>A_7</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.225000</td>\n",
       "      <td>45.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A</td>\n",
       "      <td>A_8</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>6.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id transcript_id sample  counts   pi  gene_potential  entropy  \\\n",
       "0       A           A_1  heart      50  0.5               8      1.0   \n",
       "1       A           A_2  heart       0  0.0               8      1.0   \n",
       "2       A           A_3  heart       0  0.0               8      1.0   \n",
       "3       A           A_4  heart      50  0.5               8      1.0   \n",
       "4       A           A_5  heart       0  0.0               8      1.0   \n",
       "5       A           A_6  heart       0  0.0               8      1.0   \n",
       "6       A           A_7  heart       0  0.0               8      1.0   \n",
       "7       A           A_8  heart       0  0.0               8      1.0   \n",
       "\n",
       "   perplexity  n_effective_isoforms  isoform_rank  effective_isoform  \\\n",
       "0         2.0                   2.0             1               True   \n",
       "1         2.0                   2.0             3              False   \n",
       "2         2.0                   2.0             4              False   \n",
       "3         2.0                   2.0             2               True   \n",
       "4         2.0                   2.0             5              False   \n",
       "5         2.0                   2.0             6              False   \n",
       "6         2.0                   2.0             7              False   \n",
       "7         2.0                   2.0             8              False   \n",
       "\n",
       "   n_samples_effective  expression_breadth  n_exp_samples  expression_var  \\\n",
       "0                  4.0               100.0              4        0.160078   \n",
       "1                  2.0                50.0              4        0.095743   \n",
       "2                  2.0                50.0              4        0.094516   \n",
       "3                  3.0                75.0              4        0.210000   \n",
       "4                  1.0                25.0              4        0.250000   \n",
       "5                  1.0                25.0              4        0.095219   \n",
       "6                  1.0                25.0              4        0.225000   \n",
       "7                  NaN                 0.0              4        0.030000   \n",
       "\n",
       "   avg_transcript_id_counts  avg_gene_id_counts  \n",
       "0                     33.75               100.0  \n",
       "1                     15.00               100.0  \n",
       "2                     14.00               100.0  \n",
       "3                     20.50               100.0  \n",
       "4                     50.00               100.0  \n",
       "5                     12.00               100.0  \n",
       "6                     45.00               100.0  \n",
       "7                      6.00               100.0  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['sample']=='heart']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "78c9fcd0-6a2b-4747-a8e5-0af000f980f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # temp = df[[sample_col, feature_col, 'counts']].drop_duplicates().copy(deep=True)\n",
    "\n",
    "# sample_col = 'sample'\n",
    "# feature_col = 'gene_id'\n",
    "# temp = df.copy(deep=True)\n",
    "\n",
    "# # in case the values are not unique (ie feature = gene_id, orf_id, etc. \n",
    "# temp = df[[feature_col, sample_col, 'counts']]\n",
    "# temp.head()\n",
    "# temp  = (temp.groupby([feature_col, sample_col])\n",
    "#         .sum()\n",
    "#         .reset_index()\n",
    "#         .rename({'counts':f'{feature_col}_counts'}, axis=1))\n",
    "# temp.drop(sample_col, axis=1, inplace=True)\n",
    "# temp = (temp.groupby(feature_col)\n",
    "#            .mean()\n",
    "#            .reset_index()\n",
    "#            .rename({f'{feature_col}_counts':\n",
    "#                     f'avg_{feature_col}_counts'}, axis=1))\n",
    "# df = df.merge(temp,\n",
    "#               how='left',\n",
    "#               on=[feature_col]) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "dedce66d-dc70-4f9b-8220-04056052bf11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>sample</th>\n",
       "      <th>counts</th>\n",
       "      <th>pi</th>\n",
       "      <th>gene_potential</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perplexity</th>\n",
       "      <th>n_effective_isoforms</th>\n",
       "      <th>isoform_rank</th>\n",
       "      <th>effective_isoform</th>\n",
       "      <th>n_samples_effective</th>\n",
       "      <th>expression_breadth</th>\n",
       "      <th>n_exp_samples</th>\n",
       "      <th>expression_var</th>\n",
       "      <th>avg_transcript_id_counts</th>\n",
       "      <th>avg_gene_id_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>heart</td>\n",
       "      <td>50</td>\n",
       "      <td>0.50</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.160078</td>\n",
       "      <td>33.75</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.095743</td>\n",
       "      <td>15.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>A_3</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.094516</td>\n",
       "      <td>14.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A</td>\n",
       "      <td>A_4</td>\n",
       "      <td>heart</td>\n",
       "      <td>50</td>\n",
       "      <td>0.50</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.210000</td>\n",
       "      <td>20.50</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>A_5</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>50.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A</td>\n",
       "      <td>A_6</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.095219</td>\n",
       "      <td>12.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A</td>\n",
       "      <td>A_7</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.225000</td>\n",
       "      <td>45.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A</td>\n",
       "      <td>A_8</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>6.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>brain</td>\n",
       "      <td>45</td>\n",
       "      <td>0.45</td>\n",
       "      <td>8</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.160078</td>\n",
       "      <td>33.75</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>brain</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.095743</td>\n",
       "      <td>15.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A</td>\n",
       "      <td>A_3</td>\n",
       "      <td>brain</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.094516</td>\n",
       "      <td>14.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A</td>\n",
       "      <td>A_4</td>\n",
       "      <td>brain</td>\n",
       "      <td>10</td>\n",
       "      <td>0.10</td>\n",
       "      <td>8</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.210000</td>\n",
       "      <td>20.50</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A</td>\n",
       "      <td>A_5</td>\n",
       "      <td>brain</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>50.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A</td>\n",
       "      <td>A_6</td>\n",
       "      <td>brain</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.095219</td>\n",
       "      <td>12.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A</td>\n",
       "      <td>A_7</td>\n",
       "      <td>brain</td>\n",
       "      <td>45</td>\n",
       "      <td>0.45</td>\n",
       "      <td>8</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.225000</td>\n",
       "      <td>45.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A</td>\n",
       "      <td>A_8</td>\n",
       "      <td>brain</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>6.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>lungs</td>\n",
       "      <td>20</td>\n",
       "      <td>0.20</td>\n",
       "      <td>8</td>\n",
       "      <td>2.130252</td>\n",
       "      <td>4.377939</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.160078</td>\n",
       "      <td>33.75</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>lungs</td>\n",
       "      <td>10</td>\n",
       "      <td>0.10</td>\n",
       "      <td>8</td>\n",
       "      <td>2.130252</td>\n",
       "      <td>4.377939</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.095743</td>\n",
       "      <td>15.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>A</td>\n",
       "      <td>A_3</td>\n",
       "      <td>lungs</td>\n",
       "      <td>8</td>\n",
       "      <td>0.08</td>\n",
       "      <td>8</td>\n",
       "      <td>2.130252</td>\n",
       "      <td>4.377939</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.094516</td>\n",
       "      <td>14.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A</td>\n",
       "      <td>A_4</td>\n",
       "      <td>lungs</td>\n",
       "      <td>2</td>\n",
       "      <td>0.02</td>\n",
       "      <td>8</td>\n",
       "      <td>2.130252</td>\n",
       "      <td>4.377939</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.210000</td>\n",
       "      <td>20.50</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>A</td>\n",
       "      <td>A_5</td>\n",
       "      <td>lungs</td>\n",
       "      <td>50</td>\n",
       "      <td>0.50</td>\n",
       "      <td>8</td>\n",
       "      <td>2.130252</td>\n",
       "      <td>4.377939</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>50.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>A</td>\n",
       "      <td>A_6</td>\n",
       "      <td>lungs</td>\n",
       "      <td>4</td>\n",
       "      <td>0.04</td>\n",
       "      <td>8</td>\n",
       "      <td>2.130252</td>\n",
       "      <td>4.377939</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.095219</td>\n",
       "      <td>12.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A</td>\n",
       "      <td>A_7</td>\n",
       "      <td>lungs</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8</td>\n",
       "      <td>2.130252</td>\n",
       "      <td>4.377939</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.225000</td>\n",
       "      <td>45.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>A</td>\n",
       "      <td>A_8</td>\n",
       "      <td>lungs</td>\n",
       "      <td>6</td>\n",
       "      <td>0.06</td>\n",
       "      <td>8</td>\n",
       "      <td>2.130252</td>\n",
       "      <td>4.377939</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>6.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>kidney</td>\n",
       "      <td>20</td>\n",
       "      <td>0.20</td>\n",
       "      <td>8</td>\n",
       "      <td>2.321928</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.160078</td>\n",
       "      <td>33.75</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>kidney</td>\n",
       "      <td>20</td>\n",
       "      <td>0.20</td>\n",
       "      <td>8</td>\n",
       "      <td>2.321928</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.095743</td>\n",
       "      <td>15.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A</td>\n",
       "      <td>A_3</td>\n",
       "      <td>kidney</td>\n",
       "      <td>20</td>\n",
       "      <td>0.20</td>\n",
       "      <td>8</td>\n",
       "      <td>2.321928</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.094516</td>\n",
       "      <td>14.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>A</td>\n",
       "      <td>A_4</td>\n",
       "      <td>kidney</td>\n",
       "      <td>20</td>\n",
       "      <td>0.20</td>\n",
       "      <td>8</td>\n",
       "      <td>2.321928</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.210000</td>\n",
       "      <td>20.50</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>A</td>\n",
       "      <td>A_5</td>\n",
       "      <td>kidney</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8</td>\n",
       "      <td>2.321928</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>50.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A</td>\n",
       "      <td>A_6</td>\n",
       "      <td>kidney</td>\n",
       "      <td>20</td>\n",
       "      <td>0.20</td>\n",
       "      <td>8</td>\n",
       "      <td>2.321928</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.095219</td>\n",
       "      <td>12.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>A</td>\n",
       "      <td>A_7</td>\n",
       "      <td>kidney</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8</td>\n",
       "      <td>2.321928</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.225000</td>\n",
       "      <td>45.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>A</td>\n",
       "      <td>A_8</td>\n",
       "      <td>kidney</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>8</td>\n",
       "      <td>2.321928</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>5.0</td>\n",
       "      <td>8</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.030000</td>\n",
       "      <td>6.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_id transcript_id  sample  counts    pi  gene_potential   entropy  \\\n",
       "0        A           A_1   heart      50  0.50               8  1.000000   \n",
       "1        A           A_2   heart       0  0.00               8  1.000000   \n",
       "2        A           A_3   heart       0  0.00               8  1.000000   \n",
       "3        A           A_4   heart      50  0.50               8  1.000000   \n",
       "4        A           A_5   heart       0  0.00               8  1.000000   \n",
       "5        A           A_6   heart       0  0.00               8  1.000000   \n",
       "6        A           A_7   heart       0  0.00               8  1.000000   \n",
       "7        A           A_8   heart       0  0.00               8  1.000000   \n",
       "8        A           A_1   brain      45  0.45               8  1.368996   \n",
       "9        A           A_2   brain       0  0.00               8  1.368996   \n",
       "10       A           A_3   brain       0  0.00               8  1.368996   \n",
       "11       A           A_4   brain      10  0.10               8  1.368996   \n",
       "12       A           A_5   brain       0  0.00               8  1.368996   \n",
       "13       A           A_6   brain       0  0.00               8  1.368996   \n",
       "14       A           A_7   brain      45  0.45               8  1.368996   \n",
       "15       A           A_8   brain       0  0.00               8  1.368996   \n",
       "16       A           A_1   lungs      20  0.20               8  2.130252   \n",
       "17       A           A_2   lungs      10  0.10               8  2.130252   \n",
       "18       A           A_3   lungs       8  0.08               8  2.130252   \n",
       "19       A           A_4   lungs       2  0.02               8  2.130252   \n",
       "20       A           A_5   lungs      50  0.50               8  2.130252   \n",
       "21       A           A_6   lungs       4  0.04               8  2.130252   \n",
       "22       A           A_7   lungs       0  0.00               8  2.130252   \n",
       "23       A           A_8   lungs       6  0.06               8  2.130252   \n",
       "24       A           A_1  kidney      20  0.20               8  2.321928   \n",
       "25       A           A_2  kidney      20  0.20               8  2.321928   \n",
       "26       A           A_3  kidney      20  0.20               8  2.321928   \n",
       "27       A           A_4  kidney      20  0.20               8  2.321928   \n",
       "28       A           A_5  kidney       0  0.00               8  2.321928   \n",
       "29       A           A_6  kidney      20  0.20               8  2.321928   \n",
       "30       A           A_7  kidney       0  0.00               8  2.321928   \n",
       "31       A           A_8  kidney       0  0.00               8  2.321928   \n",
       "\n",
       "    perplexity  n_effective_isoforms  isoform_rank  effective_isoform  \\\n",
       "0     2.000000                   2.0             1               True   \n",
       "1     2.000000                   2.0             3              False   \n",
       "2     2.000000                   2.0             4              False   \n",
       "3     2.000000                   2.0             2               True   \n",
       "4     2.000000                   2.0             5              False   \n",
       "5     2.000000                   2.0             6              False   \n",
       "6     2.000000                   2.0             7              False   \n",
       "7     2.000000                   2.0             8              False   \n",
       "8     2.582907                   3.0             1               True   \n",
       "9     2.582907                   3.0             4              False   \n",
       "10    2.582907                   3.0             5              False   \n",
       "11    2.582907                   3.0             3               True   \n",
       "12    2.582907                   3.0             6              False   \n",
       "13    2.582907                   3.0             7              False   \n",
       "14    2.582907                   3.0             2               True   \n",
       "15    2.582907                   3.0             8              False   \n",
       "16    4.377939                   4.0             2               True   \n",
       "17    4.377939                   4.0             3               True   \n",
       "18    4.377939                   4.0             4               True   \n",
       "19    4.377939                   4.0             7              False   \n",
       "20    4.377939                   4.0             1               True   \n",
       "21    4.377939                   4.0             6              False   \n",
       "22    4.377939                   4.0             8              False   \n",
       "23    4.377939                   4.0             5              False   \n",
       "24    5.000000                   5.0             1               True   \n",
       "25    5.000000                   5.0             2               True   \n",
       "26    5.000000                   5.0             3               True   \n",
       "27    5.000000                   5.0             4               True   \n",
       "28    5.000000                   5.0             6              False   \n",
       "29    5.000000                   5.0             5               True   \n",
       "30    5.000000                   5.0             7              False   \n",
       "31    5.000000                   5.0             8              False   \n",
       "\n",
       "    n_samples_effective  expression_breadth  n_exp_samples  expression_var  \\\n",
       "0                   4.0               100.0              4        0.160078   \n",
       "1                   2.0                50.0              4        0.095743   \n",
       "2                   2.0                50.0              4        0.094516   \n",
       "3                   3.0                75.0              4        0.210000   \n",
       "4                   1.0                25.0              4        0.250000   \n",
       "5                   1.0                25.0              4        0.095219   \n",
       "6                   1.0                25.0              4        0.225000   \n",
       "7                   NaN                 0.0              4        0.030000   \n",
       "8                   4.0               100.0              4        0.160078   \n",
       "9                   2.0                50.0              4        0.095743   \n",
       "10                  2.0                50.0              4        0.094516   \n",
       "11                  3.0                75.0              4        0.210000   \n",
       "12                  1.0                25.0              4        0.250000   \n",
       "13                  1.0                25.0              4        0.095219   \n",
       "14                  1.0                25.0              4        0.225000   \n",
       "15                  NaN                 0.0              4        0.030000   \n",
       "16                  4.0               100.0              4        0.160078   \n",
       "17                  2.0                50.0              4        0.095743   \n",
       "18                  2.0                50.0              4        0.094516   \n",
       "19                  3.0                75.0              4        0.210000   \n",
       "20                  1.0                25.0              4        0.250000   \n",
       "21                  1.0                25.0              4        0.095219   \n",
       "22                  1.0                25.0              4        0.225000   \n",
       "23                  NaN                 0.0              4        0.030000   \n",
       "24                  4.0               100.0              4        0.160078   \n",
       "25                  2.0                50.0              4        0.095743   \n",
       "26                  2.0                50.0              4        0.094516   \n",
       "27                  3.0                75.0              4        0.210000   \n",
       "28                  1.0                25.0              4        0.250000   \n",
       "29                  1.0                25.0              4        0.095219   \n",
       "30                  1.0                25.0              4        0.225000   \n",
       "31                  NaN                 0.0              4        0.030000   \n",
       "\n",
       "    avg_transcript_id_counts  avg_gene_id_counts  \n",
       "0                      33.75               100.0  \n",
       "1                      15.00               100.0  \n",
       "2                      14.00               100.0  \n",
       "3                      20.50               100.0  \n",
       "4                      50.00               100.0  \n",
       "5                      12.00               100.0  \n",
       "6                      45.00               100.0  \n",
       "7                       6.00               100.0  \n",
       "8                      33.75               100.0  \n",
       "9                      15.00               100.0  \n",
       "10                     14.00               100.0  \n",
       "11                     20.50               100.0  \n",
       "12                     50.00               100.0  \n",
       "13                     12.00               100.0  \n",
       "14                     45.00               100.0  \n",
       "15                      6.00               100.0  \n",
       "16                     33.75               100.0  \n",
       "17                     15.00               100.0  \n",
       "18                     14.00               100.0  \n",
       "19                     20.50               100.0  \n",
       "20                     50.00               100.0  \n",
       "21                     12.00               100.0  \n",
       "22                     45.00               100.0  \n",
       "23                      6.00               100.0  \n",
       "24                     33.75               100.0  \n",
       "25                     15.00               100.0  \n",
       "26                     14.00               100.0  \n",
       "27                     20.50               100.0  \n",
       "28                     50.00               100.0  \n",
       "29                     12.00               100.0  \n",
       "30                     45.00               100.0  \n",
       "31                      6.00               100.0  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7776ae3a-f8b1-452a-89b3-94d1200e2aa4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we want to be able to collapse values across feature_col / sample_col\n",
    "# to get global values\n",
    "df = dummy_counts_data_sample()\n",
    "feature_col = 'transcript_id'\n",
    "sample_col = 'sample'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "41ab4956-2481-4b5a-af0d-d94de7f0f53a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>sample</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_1</td>\n",
       "      <td>heart</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A_1</td>\n",
       "      <td>brain</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A_1</td>\n",
       "      <td>lungs</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A_1</td>\n",
       "      <td>kidney</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A_2</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  transcript_id  sample  counts\n",
       "0           A_1   heart      50\n",
       "1           A_1   brain      45\n",
       "2           A_1   lungs      20\n",
       "3           A_1  kidney      20\n",
       "4           A_2   heart       0"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df[[feature_col, sample_col, 'counts']]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3f216e4c-37fb-4180-8fce-69c8966a1380",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>sample</th>\n",
       "      <th>counts</th>\n",
       "      <th>depth</th>\n",
       "      <th>transcript_id_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A_1</td>\n",
       "      <td>heart</td>\n",
       "      <td>50</td>\n",
       "      <td>100</td>\n",
       "      <td>5.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A_1</td>\n",
       "      <td>brain</td>\n",
       "      <td>45</td>\n",
       "      <td>100</td>\n",
       "      <td>4.500000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A_1</td>\n",
       "      <td>lungs</td>\n",
       "      <td>20</td>\n",
       "      <td>100</td>\n",
       "      <td>2.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A_1</td>\n",
       "      <td>kidney</td>\n",
       "      <td>20</td>\n",
       "      <td>100</td>\n",
       "      <td>2.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A_2</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A_2</td>\n",
       "      <td>brain</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A_2</td>\n",
       "      <td>lungs</td>\n",
       "      <td>10</td>\n",
       "      <td>100</td>\n",
       "      <td>1.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A_2</td>\n",
       "      <td>kidney</td>\n",
       "      <td>20</td>\n",
       "      <td>100</td>\n",
       "      <td>2.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A_3</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A_3</td>\n",
       "      <td>brain</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A_3</td>\n",
       "      <td>lungs</td>\n",
       "      <td>8</td>\n",
       "      <td>100</td>\n",
       "      <td>8.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A_3</td>\n",
       "      <td>kidney</td>\n",
       "      <td>20</td>\n",
       "      <td>100</td>\n",
       "      <td>2.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A_4</td>\n",
       "      <td>heart</td>\n",
       "      <td>50</td>\n",
       "      <td>100</td>\n",
       "      <td>5.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A_4</td>\n",
       "      <td>brain</td>\n",
       "      <td>10</td>\n",
       "      <td>100</td>\n",
       "      <td>1.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A_4</td>\n",
       "      <td>lungs</td>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>2.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A_4</td>\n",
       "      <td>kidney</td>\n",
       "      <td>20</td>\n",
       "      <td>100</td>\n",
       "      <td>2.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A_5</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A_5</td>\n",
       "      <td>brain</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>A_5</td>\n",
       "      <td>lungs</td>\n",
       "      <td>50</td>\n",
       "      <td>100</td>\n",
       "      <td>5.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A_5</td>\n",
       "      <td>kidney</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>A_6</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>A_6</td>\n",
       "      <td>brain</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A_6</td>\n",
       "      <td>lungs</td>\n",
       "      <td>4</td>\n",
       "      <td>100</td>\n",
       "      <td>4.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>A_6</td>\n",
       "      <td>kidney</td>\n",
       "      <td>20</td>\n",
       "      <td>100</td>\n",
       "      <td>2.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A_7</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>A_7</td>\n",
       "      <td>brain</td>\n",
       "      <td>45</td>\n",
       "      <td>100</td>\n",
       "      <td>4.500000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A_7</td>\n",
       "      <td>lungs</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>A_7</td>\n",
       "      <td>kidney</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>A_8</td>\n",
       "      <td>heart</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A_8</td>\n",
       "      <td>brain</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>A_8</td>\n",
       "      <td>lungs</td>\n",
       "      <td>6</td>\n",
       "      <td>100</td>\n",
       "      <td>6.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>A_8</td>\n",
       "      <td>kidney</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   transcript_id  sample  counts  depth  transcript_id_tpm\n",
       "0            A_1   heart      50    100       5.000000e-07\n",
       "1            A_1   brain      45    100       4.500000e-07\n",
       "2            A_1   lungs      20    100       2.000000e-07\n",
       "3            A_1  kidney      20    100       2.000000e-07\n",
       "4            A_2   heart       0    100       0.000000e+00\n",
       "5            A_2   brain       0    100       0.000000e+00\n",
       "6            A_2   lungs      10    100       1.000000e-07\n",
       "7            A_2  kidney      20    100       2.000000e-07\n",
       "8            A_3   heart       0    100       0.000000e+00\n",
       "9            A_3   brain       0    100       0.000000e+00\n",
       "10           A_3   lungs       8    100       8.000000e-08\n",
       "11           A_3  kidney      20    100       2.000000e-07\n",
       "12           A_4   heart      50    100       5.000000e-07\n",
       "13           A_4   brain      10    100       1.000000e-07\n",
       "14           A_4   lungs       2    100       2.000000e-08\n",
       "15           A_4  kidney      20    100       2.000000e-07\n",
       "16           A_5   heart       0    100       0.000000e+00\n",
       "17           A_5   brain       0    100       0.000000e+00\n",
       "18           A_5   lungs      50    100       5.000000e-07\n",
       "19           A_5  kidney       0    100       0.000000e+00\n",
       "20           A_6   heart       0    100       0.000000e+00\n",
       "21           A_6   brain       0    100       0.000000e+00\n",
       "22           A_6   lungs       4    100       4.000000e-08\n",
       "23           A_6  kidney      20    100       2.000000e-07\n",
       "24           A_7   heart       0    100       0.000000e+00\n",
       "25           A_7   brain      45    100       4.500000e-07\n",
       "26           A_7   lungs       0    100       0.000000e+00\n",
       "27           A_7  kidney       0    100       0.000000e+00\n",
       "28           A_8   heart       0    100       0.000000e+00\n",
       "29           A_8   brain       0    100       0.000000e+00\n",
       "30           A_8   lungs       6    100       6.000000e-08\n",
       "31           A_8  kidney       0    100       0.000000e+00"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['depth'] = (df.groupby(sample_col)['counts']\n",
    "               .transform('sum'))\n",
    "df[f'{feature_col}_tpm'] = df['counts'] / ( df['depth'] * 1e6 )\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3b0c2f21-2ea6-4c9a-a36c-59d681cf8dff",
   "metadata": {},
   "outputs": [],
   "source": [
    "EXP_COL = 'counts'\n",
    "FEATURE_COL = TRANSCRIPT_COL = 'transcript_id'\n",
    "GENE_COL = 'gene_id'\n",
    "\n",
    "def collapse_counts_by_feature(df,\n",
    "                               feature_col=TRANSCRIPT_COL,\n",
    "                               gene_col=GENE_COL,\n",
    "                               sample_col=None):\n",
    "    \"\"\"\n",
    "    Collapse counts by a feature (e.g. ORF, transcript) instead of transcript.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        Input table with counts at transcript level.\n",
    "    feature_col : str\n",
    "        Alternative feature column to collapse to (e.g. 'orf_id').\n",
    "    gene_col : str\n",
    "        Column identifying genes.\n",
    "    sample_col : str, optional\n",
    "        Sample column. If None, assumes single-sample bulk.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        Collapsed df with summed counts per feature.\n",
    "    \"\"\"\n",
    "    group_cols = [gene_col, feature_col]\n",
    "    if sample_col is not None:\n",
    "        group_cols.append(sample_col)\n",
    "\n",
    "    # Sum counts for all transcripts mapping to the same feature\n",
    "    out = (df.groupby(group_cols, as_index=False)['counts']\n",
    "          .sum())\n",
    "\n",
    "    return out\n",
    "\n",
    "def dummy_data_with_orf():\n",
    "    \"\"\"\n",
    "    Generate dummy multi-sample isoform counts\n",
    "    with an additional ORF-level grouping.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "    \"\"\"\n",
    "    import pandas as pd\n",
    "    \n",
    "    genes = ['A', 'B']\n",
    "    samples = ['heart', 'brain', 'lungs', 'kidney']\n",
    "    \n",
    "    # map transcripts to ORFs\n",
    "    orf_map = {}\n",
    "    for g in genes:\n",
    "        orf_map[f'{g}_1'] = f'{g}_orf1'\n",
    "        orf_map[f'{g}_2'] = f'{g}_orf1'\n",
    "        orf_map[f'{g}_3'] = f'{g}_orf2'\n",
    "        orf_map[f'{g}_4'] = f'{g}_orf2'\n",
    "\n",
    "    rows = []\n",
    "    for s in samples:\n",
    "        for g in genes:\n",
    "            for i in range(1, 5):\n",
    "                t = f'{g}_{i}'\n",
    "                counts = (hash(f\"{s}_{t}\") % 50)  # deterministic dummy counts\n",
    "                rows.append({\n",
    "                    'sample': s,\n",
    "                    'gene_id': g,\n",
    "                    'transcript_id': t,\n",
    "                    'orf_id': orf_map[t],\n",
    "                    'counts': counts\n",
    "                })\n",
    "\n",
    "    return pd.DataFrame(rows)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "468c69b5-4059-4d34-8889-3bae1ebd86a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = dummy_data_with_orf()\n",
    "df2 = compute_avg_expression(df, \n",
    "                             feature_col='orf_id',\n",
    "                               sample_col='sample')\n",
    "# df2 = collapse_counts_by_feature(df,\n",
    "#                                feature_col='orf_id',\n",
    "#                                gene_col='gene_id',\n",
    "#                                sample_col='sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "b1c89b8b-36a1-45a2-8d6b-d729681b6a5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>orf_id</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>heart</td>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>A_orf1</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>heart</td>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>A_orf1</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>brain</td>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>A_orf1</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>brain</td>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>A_orf1</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>lungs</td>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>A_orf1</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>lungs</td>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>A_orf1</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>kidney</td>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>A_orf1</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>kidney</td>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>A_orf1</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample gene_id transcript_id  orf_id  counts\n",
       "0    heart       A           A_1  A_orf1      33\n",
       "1    heart       A           A_2  A_orf1      43\n",
       "8    brain       A           A_1  A_orf1      34\n",
       "9    brain       A           A_2  A_orf1      34\n",
       "16   lungs       A           A_1  A_orf1      21\n",
       "17   lungs       A           A_2  A_orf1      17\n",
       "24  kidney       A           A_1  A_orf1      30\n",
       "25  kidney       A           A_2  A_orf1      42"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.orf_id=='A_orf1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ae532398-83af-4490-b144-342211983d16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>orf_id</th>\n",
       "      <th>counts</th>\n",
       "      <th>avg_orf_id_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>heart</td>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>A_orf1</td>\n",
       "      <td>33</td>\n",
       "      <td>63.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>heart</td>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>A_orf1</td>\n",
       "      <td>43</td>\n",
       "      <td>63.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>brain</td>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>A_orf1</td>\n",
       "      <td>34</td>\n",
       "      <td>63.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>brain</td>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>A_orf1</td>\n",
       "      <td>34</td>\n",
       "      <td>63.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>lungs</td>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>A_orf1</td>\n",
       "      <td>21</td>\n",
       "      <td>63.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>lungs</td>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>A_orf1</td>\n",
       "      <td>17</td>\n",
       "      <td>63.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>kidney</td>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>A_orf1</td>\n",
       "      <td>30</td>\n",
       "      <td>63.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>kidney</td>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>A_orf1</td>\n",
       "      <td>42</td>\n",
       "      <td>63.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample gene_id transcript_id  orf_id  counts  avg_orf_id_counts\n",
       "0    heart       A           A_1  A_orf1      33               63.5\n",
       "1    heart       A           A_2  A_orf1      43               63.5\n",
       "8    brain       A           A_1  A_orf1      34               63.5\n",
       "9    brain       A           A_2  A_orf1      34               63.5\n",
       "16   lungs       A           A_1  A_orf1      21               63.5\n",
       "17   lungs       A           A_2  A_orf1      17               63.5\n",
       "24  kidney       A           A_1  A_orf1      30               63.5\n",
       "25  kidney       A           A_2  A_orf1      42               63.5"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.loc[df2.orf_id=='A_orf1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c09fa184-53af-49ae-af74-06b478d0b61e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "63.5"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "((33+43)+(34+34)+(21+17)+(30+42))/4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1b9e639-bbf3-4ad1-8f0c-72b0cff359ea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
