{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "8b80ba23-f5f8-49e4-bd1d-588cc9aa4f64",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils import *\n",
    "import scipy.stats\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy import sparse\n",
    "import h5py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "5712dae6-7f84-4f3e-89b6-cd951ccc6bd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def manuscript_sample_df():\n",
    "    df = pd.DataFrame()\n",
    "    df['gene_id'] = ['A' for i in range(8*4)]\n",
    "    df['transcript_id'] = flatten_list([[f'A_{j+1}' for i in range(4)] for j in range(8)])\n",
    "    df['sample'] = flatten_list([['heart', 'brain', 'lungs', 'kidney'] for i in range(8)])\n",
    "    df['pi'] = [i*100 for i in [0.5, 0.45, 0.2, 0.2,\n",
    "                0, 0, 0.1, 0.2,\n",
    "                0, 0, 0.08, 0.2,\n",
    "                0.5, 0.1, 0.02, 0.2,\n",
    "                0, 0, 0.5, 0,\n",
    "                0, 0, 0.04, 0.2,\n",
    "                0, 0.45, 0, 0,\n",
    "                0, 0, 0.06, 0]]\n",
    "    \n",
    "    # long formitze it\n",
    "    df = df.pivot(index=['gene_id', 'transcript_id'], columns='sample', values='pi')\n",
    "    df = df.reset_index()\n",
    "    df.columns.name = ''\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9208259d-5a85-414b-bcc5-8d8857596011",
   "metadata": {},
   "source": [
    "# remove unexpressed transcripts and genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "2839089f-8506-4a8a-b24e-c3eaab9e8ae6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def simple_counts_df():\n",
    "    \"\"\"2 genes, 2 isoforms each, with varying counts.\"\"\"\n",
    "    return pd.DataFrame({\n",
    "        \"gene_id\": [\"G1\", \"G1\", \"G2\", \"G2\"],\n",
    "        \"transcript_id\": [\"T1\", \"T2\", \"T3\", \"T4\"],\n",
    "        \"global_counts\": [100, 50, 0, 0],\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "6d7f0e0d-c6cb-43bd-8685-bf877cfdc8be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>global_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G1</td>\n",
       "      <td>T1</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G1</td>\n",
       "      <td>T2</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G2</td>\n",
       "      <td>T3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G2</td>\n",
       "      <td>T4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id transcript_id  global_counts\n",
       "0      G1            T1            100\n",
       "1      G1            T2             50\n",
       "2      G2            T3              0\n",
       "3      G2            T4              0"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = simple_counts_df()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "a49e6720-0b94-4678-bdfe-a98435044e49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G1</td>\n",
       "      <td>T1</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G1</td>\n",
       "      <td>T2</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G2</td>\n",
       "      <td>T3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G2</td>\n",
       "      <td>T4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id transcript_id  counts\n",
       "0      G1            T1     100\n",
       "1      G1            T2      50\n",
       "2      G2            T3       0\n",
       "3      G2            T4       0"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "gene_col = 'gene_id'\n",
    "feature_col = 'transcript_id'\n",
    "expression_type = 'counts'\n",
    "\n",
    "validate_counts_input(df,\n",
    "                          gene_col=gene_col,\n",
    "                          feature_col=feature_col)\n",
    "df = rename_sample_col(df,\n",
    "                  gene_col=gene_col,\n",
    "                  feature_col=feature_col,\n",
    "                  expression_type=expression_type)\n",
    "\n",
    "# collapse counts if feature_col is not unique\n",
    "df = collapse_counts_by_feature(df,\n",
    "                                feature_col=feature_col,\n",
    "                                expression_type=expression_type,\n",
    "                                gene_col=gene_col,\n",
    "                                sample_col=None)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "c7bc2237-d665-4366-9d52-4e1360502421",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G1</td>\n",
       "      <td>T1</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G1</td>\n",
       "      <td>T2</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id transcript_id  counts\n",
       "0      G1            T1     100\n",
       "1      G1            T2      50"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove unexpressed transcripts \n",
    "df = df.loc[df.counts>0]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19f13860-b4f0-4151-bcb1-bfe63fbf2247",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d51f0aa4-79d3-42b3-8d89-eed0d2de08e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>brain</th>\n",
       "      <th>heart</th>\n",
       "      <th>kidney</th>\n",
       "      <th>lungs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>45.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>A_3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A</td>\n",
       "      <td>A_4</td>\n",
       "      <td>10.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>A_5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id transcript_id  brain  heart  kidney  lungs\n",
       "0       A           A_1   45.0   50.0    20.0   20.0\n",
       "1       A           A_2    0.0    0.0    20.0   10.0\n",
       "2       A           A_3    0.0    0.0    20.0    8.0\n",
       "3       A           A_4   10.0   50.0    20.0    2.0\n",
       "4       A           A_5    0.0    0.0     0.0   50.0"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = manuscript_sample_df()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7ed78a3f-afc4-4838-9090-cde5ef782737",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_col = 'gene_id'\n",
    "feature_col = 'transcript_id'\n",
    "expression_type = 'counts'\n",
    "sample_df, global_df = compute_multi_sample_isoform_metrics(df, gene_col=gene_col,\n",
    "                                          feature_col=feature_col,\n",
    "                                          expression_type=expression_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2c5242e8-ee78-4de8-bdda-360015410deb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>counts</th>\n",
       "      <th>tpm</th>\n",
       "      <th>pi</th>\n",
       "      <th>n_detected_features</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perplexity</th>\n",
       "      <th>n_effective</th>\n",
       "      <th>feature_rank</th>\n",
       "      <th>effective</th>\n",
       "      <th>sample</th>\n",
       "      <th>n_samples_effective</th>\n",
       "      <th>expression_breadth</th>\n",
       "      <th>n_exp_samples</th>\n",
       "      <th>expression_var</th>\n",
       "      <th>avg_transcript_id_tpm</th>\n",
       "      <th>max_transcript_id_tpm</th>\n",
       "      <th>avg_gene_id_tpm</th>\n",
       "      <th>max_gene_id_tpm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "      <td>45.0</td>\n",
       "      <td>450000.0</td>\n",
       "      <td>0.45</td>\n",
       "      <td>3</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>brain</td>\n",
       "      <td>4.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.160078</td>\n",
       "      <td>337500.0</td>\n",
       "      <td>500000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "      <td>brain</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.095743</td>\n",
       "      <td>150000.0</td>\n",
       "      <td>200000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>A_3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5</td>\n",
       "      <td>False</td>\n",
       "      <td>brain</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.094516</td>\n",
       "      <td>140000.0</td>\n",
       "      <td>200000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A</td>\n",
       "      <td>A_4</td>\n",
       "      <td>10.0</td>\n",
       "      <td>100000.0</td>\n",
       "      <td>0.10</td>\n",
       "      <td>3</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>brain</td>\n",
       "      <td>3.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.210000</td>\n",
       "      <td>205000.0</td>\n",
       "      <td>500000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>A_5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3</td>\n",
       "      <td>1.368996</td>\n",
       "      <td>2.582907</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6</td>\n",
       "      <td>False</td>\n",
       "      <td>brain</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>500000.0</td>\n",
       "      <td>500000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "      <td>1000000.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id transcript_id  counts       tpm    pi  n_detected_features  \\\n",
       "0       A           A_1    45.0  450000.0  0.45                    3   \n",
       "1       A           A_2     0.0       0.0  0.00                    3   \n",
       "2       A           A_3     0.0       0.0  0.00                    3   \n",
       "3       A           A_4    10.0  100000.0  0.10                    3   \n",
       "4       A           A_5     0.0       0.0  0.00                    3   \n",
       "\n",
       "    entropy  perplexity  n_effective  feature_rank  effective sample  \\\n",
       "0  1.368996    2.582907          3.0             1       True  brain   \n",
       "1  1.368996    2.582907          3.0             4      False  brain   \n",
       "2  1.368996    2.582907          3.0             5      False  brain   \n",
       "3  1.368996    2.582907          3.0             3       True  brain   \n",
       "4  1.368996    2.582907          3.0             6      False  brain   \n",
       "\n",
       "   n_samples_effective  expression_breadth  n_exp_samples  expression_var  \\\n",
       "0                  4.0               100.0              4        0.160078   \n",
       "1                  2.0                50.0              4        0.095743   \n",
       "2                  2.0                50.0              4        0.094516   \n",
       "3                  3.0                75.0              4        0.210000   \n",
       "4                  1.0                25.0              4        0.250000   \n",
       "\n",
       "   avg_transcript_id_tpm  max_transcript_id_tpm  avg_gene_id_tpm  \\\n",
       "0               337500.0               500000.0        1000000.0   \n",
       "1               150000.0               200000.0        1000000.0   \n",
       "2               140000.0               200000.0        1000000.0   \n",
       "3               205000.0               500000.0        1000000.0   \n",
       "4               500000.0               500000.0        1000000.0   \n",
       "\n",
       "   max_gene_id_tpm  \n",
       "0        1000000.0  \n",
       "1        1000000.0  \n",
       "2        1000000.0  \n",
       "3        1000000.0  \n",
       "4        1000000.0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "f1e1ba7d-5458-44a4-9f64-7e17b7d7b959",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "b51a9413-fa0e-4b25-bc92-84ff837f4658",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "effective\n",
      "tpm\n",
      "pi\n",
      "feature_rank\n",
      "counts\n"
     ]
    }
   ],
   "source": [
    "feature_out = 'test_feature.h5'\n",
    "\n",
    "def create_h5_sparse_table(df, fname, mode, key):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    with h5py.File(fname, mode) as f:\n",
    "        g = f.create_group(key)\n",
    "        \n",
    "        # get all of these separately\n",
    "        X = sparse.csr_matrix(df.values)\n",
    "        obs_ids = np.array(df.columns)\n",
    "        var_ids = np.array(df.index)\n",
    "        \n",
    "        # add these all as hierarchical keys\n",
    "        g.create_dataset('data', data=X.data)\n",
    "        g.create_dataset('indices', data=X.indices)\n",
    "        g.create_dataset('indptr', data=X.indptr)\n",
    "        g.create_dataset('shape', data=X.shape)\n",
    "        g.create_dataset('obs', data=obs_ids.astype('S'))\n",
    "        g.create_dataset('var', data=var_ids.astype('S'))\n",
    "\n",
    "# first, make the isoform stack of DFs\n",
    "feature_cols = list(set([expression_type, 'tpm', 'pi', 'feature_rank', 'effective']))\n",
    "feature_ids = [feature_col, 'sample']\n",
    "for i, c in enumerate(feature_cols):\n",
    "    print(c)\n",
    "    temp = sample_df[[c]+feature_ids]\n",
    "    temp = temp.pivot(index=feature_col, columns='sample', values=c)\n",
    "    temp.columns.name = ''\n",
    "    \n",
    "    if i == 0: mode = 'w' \n",
    "    else: mode = 'a'\n",
    "    \n",
    "    key = f'feature/{c}'\n",
    "    create_h5_sparse_table(temp, feature_out, mode, key)    \n",
    "    \n",
    "# # with pd.HDFStore(feature_out, \"r\") as store:\n",
    "# #     print(store.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "dbfa302b-03d4-46ea-8008-bad303a3e227",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>effective</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>True</td>\n",
       "      <td>A_1</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>A_2</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>False</td>\n",
       "      <td>A_3</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>A_4</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>False</td>\n",
       "      <td>A_5</td>\n",
       "      <td>brain</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   effective transcript_id sample\n",
       "0       True           A_1  brain\n",
       "1      False           A_2  brain\n",
       "2      False           A_3  brain\n",
       "3       True           A_4  brain\n",
       "4      False           A_5  brain"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "7b6a0497-be73-4cb4-8968-bef745a3edbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>brain</th>\n",
       "      <th>heart</th>\n",
       "      <th>kidney</th>\n",
       "      <th>lungs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A_1</th>\n",
       "      <td>0.45</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_2</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_3</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_4</th>\n",
       "      <td>0.10</td>\n",
       "      <td>0.5</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A_5</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               brain  heart  kidney  lungs\n",
       "transcript_id                             \n",
       "A_1             0.45    0.5     0.2   0.20\n",
       "A_2             0.00    0.0     0.2   0.10\n",
       "A_3             0.00    0.0     0.2   0.08\n",
       "A_4             0.10    0.5     0.2   0.02\n",
       "A_5             0.00    0.0     0.0   0.50"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = pd.read_hdf(feature_out, 'pi')\n",
    "temp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a1cd2208-c0d1-4296-9f7f-af0898f67d80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RangeIndex(start=0, stop=8, step=1)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "9fc82b4d-c34d-4edc-b452-489a4f42af36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>A_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>A_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>A_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A</td>\n",
       "      <td>A_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A</td>\n",
       "      <td>A_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A</td>\n",
       "      <td>A_6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A</td>\n",
       "      <td>A_7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A</td>\n",
       "      <td>A_8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id transcript_id\n",
       "0       A           A_1\n",
       "1       A           A_2\n",
       "2       A           A_3\n",
       "3       A           A_4\n",
       "4       A           A_5\n",
       "5       A           A_6\n",
       "6       A           A_7\n",
       "7       A           A_8"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62821c99-a141-4dac-8377-0068d6160cba",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
